{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "collapsed_sections": [
        "HRQ_wRC8560m"
      ],
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    },
    "widgets": {
      "application/vnd.jupyter.widget-state+json": {
        "0a50d17db302436086d575423938a570": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "TextModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "TextModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "TextView",
            "continuous_update": true,
            "description": "Input:",
            "description_tooltip": null,
            "disabled": false,
            "layout": "IPY_MODEL_85210f49294149fabd66ec9f27d4076e",
            "placeholder": "e.g. 6  K/M",
            "style": "IPY_MODEL_f58043a3bbe14371a997d86d5d6f1bf8",
            "value": "435 L/F"
          }
        },
        "85210f49294149fabd66ec9f27d4076e": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "f58043a3bbe14371a997d86d5d6f1bf8": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "DescriptionStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        }
      }
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/glevans/7ADD-workshop-2024/blob/main/variants_embl-ebi_july2025/Example_1_structural_assessment.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "**<H1>Structural assessment of a variant</H1>**\n",
        "**Using PDBe-KB & 3D-Beacons to gain understanding on genetic variants**\n",
        "<img src=\"https://www.ebi.ac.uk/pdbe/docs_dev/logos/images/RGB/PDBe-logo-RGB_2013.png\" height=\"300\" align=\"right\">\n",
        "\n",
        "#Welcome to this notebook form!\n",
        "\n",
        "To use this notebook form you can:\n",
        "\n",
        "(1) Click on link at top of the page when viewing the file in Github\n",
        "\n",
        "*   you will need to have a Google account\n",
        "*   be logged in to Google Colab <br>(by being logged into Google account)\n",
        "\n",
        "<br>\n",
        "\n",
        "(2) Visit the Colaboratory page:<br>\n",
        "https://colab.research.google.com/<br>\n",
        "and access the following Github repository *via* the interface:<br>\n",
        "https://github.com/PDBeurope/pdbe-notebooks/\n",
        "\n",
        "<br>\n",
        "\n",
        "(3) Visit the Colaboratory page:<br>\n",
        "https://colab.research.google.com/<br>\n",
        "and upload the file from the Colaboratory interface\n",
        "\n",
        "*   you will need to have a Google account\n",
        "*   be logged in to Google Colab <br>(by being logged into Google account)"
      ],
      "metadata": {
        "id": "MEhhtOcBfwxf"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "\n",
        "\n",
        "---\n",
        "\n",
        "  ## How to use this notebook form:\n",
        "1. To run a code cell, click on the cell to select it. You will notice a play button (▶️) on the left side of the cell. Click on the play button or **press Shift+Enter** to run the code in the selected cell.\n",
        "2. The code will start executing, and you will see the output, if any, displayed below the code cell.\n",
        "3. Move to the next code cell and repeat steps 2 and 3 until you have executed all the desired code cells in sequence.\n",
        "4. The currently running step is indicated by a circle with a stop sign next to it.\n",
        "If you need to stop or interrupt the execution of a code cell, you can click on the stop button (■) located next to the play button.\n",
        "5. The code behind the form can be hidden by selecting View → Show/hide code or using the toolbar above the selected code cell.\n",
        "6. The cells in this notebook primarily are forms with pulldown options. In a few case they run small bits of code to generate a visual.\n",
        "\n",
        "*Remember to run the code cells in the correct order, as their execution might depend on variables or functions defined in previous cells. You can modify the code in a code cell and re-run it to see updated results.*\n",
        "\n",
        "<br>\n",
        "\n",
        "---\n",
        "\n",
        "## Contact us\n",
        "\n",
        "If you experience any bugs please contact pdbehelp@ebi.ac.uk and put \"Help with\" and the title of the notebook form in the subject line of the message.\n"
      ],
      "metadata": {
        "id": "6Mg5aemEWkiz"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Example 1"
      ],
      "metadata": {
        "id": "ljhELtyH2WUf"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "**Presenilin-1** is a protein component/subunit in a protein complex called the gamma secretase complex.\n",
        "\n",
        "<br>\n",
        "\n",
        "A feature from the protein perspective is: <br>\n",
        "InterPro Protein Family **Presenilin/signal peptide peptidase** ID [IPR006639](https://www.ebi.ac.uk/interpro/entry/InterPro/IPR006639/protein/reviewed/#table).\n",
        "\n",
        "*   Gene **PSEN1** encodes protein **Presenilin-1** (UniProt ID: [P49768](https://www.uniprot.org/uniprotkb/P49768/entry))\n",
        "*   Gene **PSEN2** encodes protein **Presenilin-2** (UniProt ID: [P49810](https://www.uniprot.org/uniprotkb/P49810/entry))\n",
        "\n",
        "<br>\n",
        "\n",
        "In this notebook we will focus on: <br>\n",
        "**Presenilin-1** <br>(UniProt ID: P49768)"
      ],
      "metadata": {
        "id": "dosRTZdY2YNM"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "By examining the subpages of [PDBe-KB.org](PDBe-KB.org), we can simplify possibly causes that make a variant <i>likely pathogenic or pathogenic</i>.\n",
        "\n",
        "---\n",
        "\n",
        "*   **Ligands**\n",
        "*   **Interactions**"
      ],
      "metadata": {
        "id": "9Phhgvt09WHj"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# @title Initial Questions\n",
        "# @markdown\n",
        "# @markdown\n",
        "UniProt_ID = 'P49768' # @param {type:\"string\"}\n",
        "# @markdown\n",
        "# @markdown ---\n",
        "# @markdown Do we have 3D-structural experimental data for ligands as potential reasons for a variant's pathogenicity?\n",
        "Ligands_pos = 'Yes' # @param [\"Input Answer\",\"Yes\", \"No\"]\n",
        "# @markdown ---\n",
        "# @markdown Do we have 3D-structural experimental data for metals as potential reasons for a variant's pathogenicity?\n",
        "Metals_pos = 'No' # @param [\"Input Answer\",\"Yes\", \"No\"]\n",
        "# @markdown ---\n",
        "# @markdown Do we have 3D-structural experimental data for complexes as potential reasons for a variant's pathogenicity?\n",
        "Complexes_pos = 'Yes' # @param [\"Input Answer\",\"Yes\", \"No\"]\n",
        "# @markdown ---"
      ],
      "metadata": {
        "cellView": "form",
        "id": "qI26uZKx-lUJ"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Setup"
      ],
      "metadata": {
        "id": "vuxJB8xWO_-G"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# @title **Press shift + enter to install packages**\n",
        "\n",
        "import io\n",
        "from IPython.display import display\n",
        "\n",
        "!pip install --upgrade pip\n",
        "\n",
        "!pip install ipywidgets\n",
        "import ipywidgets as widgets\n",
        "\n",
        "!pip install ColabTurtlePlus\n",
        "from ColabTurtlePlus.Turtle import *\n",
        "\n",
        "print()\n",
        "print(\"Succesfully installed!\")"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "cellView": "form",
        "id": "GhSYq90-5lfc",
        "outputId": "1d0db590-2284-4c77-c114-63b008648469"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Requirement already satisfied: pip in /usr/local/lib/python3.11/dist-packages (24.1.2)\n",
            "Collecting pip\n",
            "  Downloading pip-25.1.1-py3-none-any.whl.metadata (3.6 kB)\n",
            "Downloading pip-25.1.1-py3-none-any.whl (1.8 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m1.8/1.8 MB\u001b[0m \u001b[31m21.0 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hInstalling collected packages: pip\n",
            "  Attempting uninstall: pip\n",
            "    Found existing installation: pip 24.1.2\n",
            "    Uninstalling pip-24.1.2:\n",
            "      Successfully uninstalled pip-24.1.2\n",
            "Successfully installed pip-25.1.1\n",
            "Requirement already satisfied: ipywidgets in /usr/local/lib/python3.11/dist-packages (7.7.1)\n",
            "Requirement already satisfied: ipykernel>=4.5.1 in /usr/local/lib/python3.11/dist-packages (from ipywidgets) (6.17.1)\n",
            "Requirement already satisfied: ipython-genutils~=0.2.0 in /usr/local/lib/python3.11/dist-packages (from ipywidgets) (0.2.0)\n",
            "Requirement already satisfied: traitlets>=4.3.1 in /usr/local/lib/python3.11/dist-packages (from ipywidgets) (5.7.1)\n",
            "Requirement already satisfied: widgetsnbextension~=3.6.0 in /usr/local/lib/python3.11/dist-packages (from ipywidgets) (3.6.10)\n",
            "Requirement already satisfied: ipython>=4.0.0 in /usr/local/lib/python3.11/dist-packages (from ipywidgets) (7.34.0)\n",
            "Requirement already satisfied: jupyterlab-widgets>=1.0.0 in /usr/local/lib/python3.11/dist-packages (from ipywidgets) (3.0.15)\n",
            "Requirement already satisfied: notebook>=4.4.1 in /usr/local/lib/python3.11/dist-packages (from widgetsnbextension~=3.6.0->ipywidgets) (6.5.7)\n",
            "Requirement already satisfied: debugpy>=1.0 in /usr/local/lib/python3.11/dist-packages (from ipykernel>=4.5.1->ipywidgets) (1.8.0)\n",
            "Requirement already satisfied: jupyter-client>=6.1.12 in /usr/local/lib/python3.11/dist-packages (from ipykernel>=4.5.1->ipywidgets) (6.1.12)\n",
            "Requirement already satisfied: matplotlib-inline>=0.1 in /usr/local/lib/python3.11/dist-packages (from ipykernel>=4.5.1->ipywidgets) (0.1.7)\n",
            "Requirement already satisfied: nest-asyncio in /usr/local/lib/python3.11/dist-packages (from ipykernel>=4.5.1->ipywidgets) (1.6.0)\n",
            "Requirement already satisfied: packaging in /usr/local/lib/python3.11/dist-packages (from ipykernel>=4.5.1->ipywidgets) (25.0)\n",
            "Requirement already satisfied: psutil in /usr/local/lib/python3.11/dist-packages (from ipykernel>=4.5.1->ipywidgets) (5.9.5)\n",
            "Requirement already satisfied: pyzmq>=17 in /usr/local/lib/python3.11/dist-packages (from ipykernel>=4.5.1->ipywidgets) (24.0.1)\n",
            "Requirement already satisfied: tornado>=6.1 in /usr/local/lib/python3.11/dist-packages (from ipykernel>=4.5.1->ipywidgets) (6.4.2)\n",
            "Requirement already satisfied: setuptools>=18.5 in /usr/local/lib/python3.11/dist-packages (from ipython>=4.0.0->ipywidgets) (75.2.0)\n",
            "Collecting jedi>=0.16 (from ipython>=4.0.0->ipywidgets)\n",
            "  Downloading jedi-0.19.2-py2.py3-none-any.whl.metadata (22 kB)\n",
            "Requirement already satisfied: decorator in /usr/local/lib/python3.11/dist-packages (from ipython>=4.0.0->ipywidgets) (4.4.2)\n",
            "Requirement already satisfied: pickleshare in /usr/local/lib/python3.11/dist-packages (from ipython>=4.0.0->ipywidgets) (0.7.5)\n",
            "Requirement already satisfied: prompt-toolkit!=3.0.0,!=3.0.1,<3.1.0,>=2.0.0 in /usr/local/lib/python3.11/dist-packages (from ipython>=4.0.0->ipywidgets) (3.0.51)\n",
            "Requirement already satisfied: pygments in /usr/local/lib/python3.11/dist-packages (from ipython>=4.0.0->ipywidgets) (2.19.2)\n",
            "Requirement already satisfied: backcall in /usr/local/lib/python3.11/dist-packages (from ipython>=4.0.0->ipywidgets) (0.2.0)\n",
            "Requirement already satisfied: pexpect>4.3 in /usr/local/lib/python3.11/dist-packages (from ipython>=4.0.0->ipywidgets) (4.9.0)\n",
            "Requirement already satisfied: wcwidth in /usr/local/lib/python3.11/dist-packages (from prompt-toolkit!=3.0.0,!=3.0.1,<3.1.0,>=2.0.0->ipython>=4.0.0->ipywidgets) (0.2.13)\n",
            "Requirement already satisfied: parso<0.9.0,>=0.8.4 in /usr/local/lib/python3.11/dist-packages (from jedi>=0.16->ipython>=4.0.0->ipywidgets) (0.8.4)\n",
            "Requirement already satisfied: jupyter-core>=4.6.0 in /usr/local/lib/python3.11/dist-packages (from jupyter-client>=6.1.12->ipykernel>=4.5.1->ipywidgets) (5.8.1)\n",
            "Requirement already satisfied: python-dateutil>=2.1 in /usr/local/lib/python3.11/dist-packages (from jupyter-client>=6.1.12->ipykernel>=4.5.1->ipywidgets) (2.9.0.post0)\n",
            "Requirement already satisfied: platformdirs>=2.5 in /usr/local/lib/python3.11/dist-packages (from jupyter-core>=4.6.0->jupyter-client>=6.1.12->ipykernel>=4.5.1->ipywidgets) (4.3.8)\n",
            "Requirement already satisfied: jinja2 in /usr/local/lib/python3.11/dist-packages (from notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets) (3.1.6)\n",
            "Requirement already satisfied: argon2-cffi in /usr/local/lib/python3.11/dist-packages (from notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets) (25.1.0)\n",
            "Requirement already satisfied: nbformat in /usr/local/lib/python3.11/dist-packages (from notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets) (5.10.4)\n",
            "Requirement already satisfied: nbconvert>=5 in /usr/local/lib/python3.11/dist-packages (from notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets) (7.16.6)\n",
            "Requirement already satisfied: Send2Trash>=1.8.0 in /usr/local/lib/python3.11/dist-packages (from notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets) (1.8.3)\n",
            "Requirement already satisfied: terminado>=0.8.3 in /usr/local/lib/python3.11/dist-packages (from notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets) (0.18.1)\n",
            "Requirement already satisfied: prometheus-client in /usr/local/lib/python3.11/dist-packages (from notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets) (0.22.1)\n",
            "Requirement already satisfied: nbclassic>=0.4.7 in /usr/local/lib/python3.11/dist-packages (from notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets) (1.3.1)\n",
            "Requirement already satisfied: notebook-shim>=0.2.3 in /usr/local/lib/python3.11/dist-packages (from nbclassic>=0.4.7->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets) (0.2.4)\n",
            "Requirement already satisfied: beautifulsoup4 in /usr/local/lib/python3.11/dist-packages (from nbconvert>=5->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets) (4.13.4)\n",
            "Requirement already satisfied: bleach!=5.0.0 in /usr/local/lib/python3.11/dist-packages (from bleach[css]!=5.0.0->nbconvert>=5->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets) (6.2.0)\n",
            "Requirement already satisfied: defusedxml in /usr/local/lib/python3.11/dist-packages (from nbconvert>=5->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets) (0.7.1)\n",
            "Requirement already satisfied: jupyterlab-pygments in /usr/local/lib/python3.11/dist-packages (from nbconvert>=5->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets) (0.3.0)\n",
            "Requirement already satisfied: markupsafe>=2.0 in /usr/local/lib/python3.11/dist-packages (from nbconvert>=5->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets) (3.0.2)\n",
            "Requirement already satisfied: mistune<4,>=2.0.3 in /usr/local/lib/python3.11/dist-packages (from nbconvert>=5->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets) (3.1.3)\n",
            "Requirement already satisfied: nbclient>=0.5.0 in /usr/local/lib/python3.11/dist-packages (from nbconvert>=5->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets) (0.10.2)\n",
            "Requirement already satisfied: pandocfilters>=1.4.1 in /usr/local/lib/python3.11/dist-packages (from nbconvert>=5->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets) (1.5.1)\n",
            "Requirement already satisfied: webencodings in /usr/local/lib/python3.11/dist-packages (from bleach!=5.0.0->bleach[css]!=5.0.0->nbconvert>=5->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets) (0.5.1)\n",
            "Requirement already satisfied: tinycss2<1.5,>=1.1.0 in /usr/local/lib/python3.11/dist-packages (from bleach[css]!=5.0.0->nbconvert>=5->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets) (1.4.0)\n",
            "Requirement already satisfied: fastjsonschema>=2.15 in /usr/local/lib/python3.11/dist-packages (from nbformat->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets) (2.21.1)\n",
            "Requirement already satisfied: jsonschema>=2.6 in /usr/local/lib/python3.11/dist-packages (from nbformat->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets) (4.24.0)\n",
            "Requirement already satisfied: attrs>=22.2.0 in /usr/local/lib/python3.11/dist-packages (from jsonschema>=2.6->nbformat->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets) (25.3.0)\n",
            "Requirement already satisfied: jsonschema-specifications>=2023.03.6 in /usr/local/lib/python3.11/dist-packages (from jsonschema>=2.6->nbformat->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets) (2025.4.1)\n",
            "Requirement already satisfied: referencing>=0.28.4 in /usr/local/lib/python3.11/dist-packages (from jsonschema>=2.6->nbformat->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets) (0.36.2)\n",
            "Requirement already satisfied: rpds-py>=0.7.1 in /usr/local/lib/python3.11/dist-packages (from jsonschema>=2.6->nbformat->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets) (0.26.0)\n",
            "Requirement already satisfied: jupyter-server<3,>=1.8 in /usr/local/lib/python3.11/dist-packages (from notebook-shim>=0.2.3->nbclassic>=0.4.7->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets) (1.16.0)\n",
            "Requirement already satisfied: anyio>=3.1.0 in /usr/local/lib/python3.11/dist-packages (from jupyter-server<3,>=1.8->notebook-shim>=0.2.3->nbclassic>=0.4.7->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets) (4.9.0)\n",
            "Requirement already satisfied: websocket-client in /usr/local/lib/python3.11/dist-packages (from jupyter-server<3,>=1.8->notebook-shim>=0.2.3->nbclassic>=0.4.7->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets) (1.8.0)\n",
            "Requirement already satisfied: idna>=2.8 in /usr/local/lib/python3.11/dist-packages (from anyio>=3.1.0->jupyter-server<3,>=1.8->notebook-shim>=0.2.3->nbclassic>=0.4.7->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets) (3.10)\n",
            "Requirement already satisfied: sniffio>=1.1 in /usr/local/lib/python3.11/dist-packages (from anyio>=3.1.0->jupyter-server<3,>=1.8->notebook-shim>=0.2.3->nbclassic>=0.4.7->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets) (1.3.1)\n",
            "Requirement already satisfied: typing_extensions>=4.5 in /usr/local/lib/python3.11/dist-packages (from anyio>=3.1.0->jupyter-server<3,>=1.8->notebook-shim>=0.2.3->nbclassic>=0.4.7->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets) (4.14.1)\n",
            "Requirement already satisfied: ptyprocess>=0.5 in /usr/local/lib/python3.11/dist-packages (from pexpect>4.3->ipython>=4.0.0->ipywidgets) (0.7.0)\n",
            "Requirement already satisfied: six>=1.5 in /usr/local/lib/python3.11/dist-packages (from python-dateutil>=2.1->jupyter-client>=6.1.12->ipykernel>=4.5.1->ipywidgets) (1.17.0)\n",
            "Requirement already satisfied: argon2-cffi-bindings in /usr/local/lib/python3.11/dist-packages (from argon2-cffi->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets) (21.2.0)\n",
            "Requirement already satisfied: cffi>=1.0.1 in /usr/local/lib/python3.11/dist-packages (from argon2-cffi-bindings->argon2-cffi->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets) (1.17.1)\n",
            "Requirement already satisfied: pycparser in /usr/local/lib/python3.11/dist-packages (from cffi>=1.0.1->argon2-cffi-bindings->argon2-cffi->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets) (2.22)\n",
            "Requirement already satisfied: soupsieve>1.2 in /usr/local/lib/python3.11/dist-packages (from beautifulsoup4->nbconvert>=5->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets) (2.7)\n",
            "Downloading jedi-0.19.2-py2.py3-none-any.whl (1.6 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m1.6/1.6 MB\u001b[0m \u001b[31m31.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hInstalling collected packages: jedi\n",
            "Successfully installed jedi-0.19.2\n",
            "Collecting ColabTurtlePlus\n",
            "  Downloading ColabTurtlePlus-2.0.1-py3-none-any.whl.metadata (10 kB)\n",
            "Downloading ColabTurtlePlus-2.0.1-py3-none-any.whl (31 kB)\n",
            "Installing collected packages: ColabTurtlePlus\n",
            "Successfully installed ColabTurtlePlus-2.0.1\n",
            "Put clearscreen() as the first line in a cell (after the import command) to re-run turtle commands in the cell\n",
            "\n",
            "Succesfully installed!\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# @title **Press shift + enter to install more packages**\n",
        "\n",
        "import pandas as pd\n",
        "import json\n",
        "import csv\n",
        "import os\n",
        "\n",
        "print(\"Succesfully installed!\")"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "cellView": "form",
        "id": "HTO2rv6HPJZj",
        "outputId": "de88ae6a-0375-4399-e17c-1f53227be0c8"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Succesfully installed!\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# 3D structures\n",
        "\n",
        "There are various ways to obtain the below information.\n",
        "\n",
        "We have another Notebook, the [Structures Available Notebook](https://colab.research.google.com/github/PDBeurope/pdbe-notebooks/blob/main/variants_embl-ebi_may2024/structures_available.ipynb), to aid in finding the information quickly. For this activity section 3.4 is the most useful.\n",
        "\n",
        "Alternatively, a combination of querying and examining contents on the webpages of [3D Beacons](https://www.ebi.ac.uk/pdbe/pdbe-kb/3dbeacons/) and [PDBe-KB](https://www.ebi.ac.uk/pdbe/pdbe-kb/) should enable the same information to be obtained."
      ],
      "metadata": {
        "id": "uMKTDiW8GAJT"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# @title **Structural data availability**\n",
        "\n",
        "UniProt_ID = 'P49768' # @param {type:\"string\"}\n",
        "# @markdown\n",
        "# @markdown ---\n",
        "# @markdown How many **structures (both predicted and experimentally-determined)** are available for this UniProt ID?\n",
        "No_of_all_available = 34 # @param {type:\"number\"}\n",
        "# @markdown HINT: Use **3D-Beacons** query\n",
        "# @markdown\n",
        "# @markdown ---\n",
        "# @markdown How many **predicted structures** (from *AlphaFoldDB*, *Swiss-Model*, *ModelArchive*, etc) are available for this UniProt ID?\n",
        "No_of_predicted_available = 8 # @param {type:\"number\"}\n",
        "# @markdown HINT: Use **3D-Beacons** query\n",
        "# @markdown\n",
        "# @markdown ---\n",
        "# @markdown How many **experimentally-determined structures** (from *PDB* & *Small angle scattering Biological Data Bank (SASBDB)*) are available for this UniProt ID?\n",
        "No_of_exp_available = 26 # @param {type:\"number\"}\n",
        "# @markdown HINT: Use either **3D-Beacons**\n",
        "# @markdown\n",
        "# @markdown ---\n",
        "# @markdown How many **experimentally-determined structures (with PDB ids)** are available for this UniProt ID?\n",
        "No_of_PDB_available = 26 # @param {type:\"number\"}\n",
        "# @markdown HINT: Use either **3D-Beacons** or **PDBe-KB**\n",
        "# @markdown\n",
        "# @markdown ---\n",
        "# @markdown How many experimentally-determined PDB structures are available for this UniProt ID with a **resolution better than 3 Ångstrom**?\n",
        "No_of_3A_and_better_available = 14 # @param {type:\"number\"}\n",
        "# @markdown HINT: Use **PDBe-KB**\n",
        "# @markdown\n",
        "\n",
        "if No_of_exp_available > 0:\n",
        "  EXPERIMENTALLY_DETERMINED_STRUCTURES = 'Yes'\n",
        "else:\n",
        "  EXPERIMENTALLY_DETERMINED_STRUCTURES = 'No'\n",
        "\n",
        "# Print the summary\n",
        "print(\"The following is a summary based on what has been filled-in so far.\")\n",
        "print()\n",
        "print(f\"There are {No_of_all_available} structures (predicted and experimentally determined) available associated with UniProt ID {UniProt_ID}.\")\n",
        "if EXPERIMENTALLY_DETERMINED_STRUCTURES == 'Yes':\n",
        "  print(f\"There are {No_of_exp_available} experimentally-determined structures associated with UniProt ID {UniProt_ID}.\")\n",
        "  print(f\"There are {No_of_PDB_available} experimentally-determined PDB structures associated with UniProt ID {UniProt_ID}.\")\n",
        "else:\n",
        "  print(f\"There are no experimentalLY-determined structures available.\")\n",
        "print()\n",
        "\n",
        "# Initialize an empty list to store the summaries\n",
        "summaries = []\n",
        "\n",
        "## Define the file to check\n",
        "#file_to_check = \"/content/struct_avail.csv\"\n",
        "## Loop while the file exists\n",
        "#if os.path.exists(file_to_check):\n",
        "#    print(f\"File '{file_to_check}' exists.\")\n",
        "#    # Load the CSV file into a DataFrame\n",
        "#    df = pd.read_csv(file_to_check)\n",
        "#    # Get the number of rows\n",
        "#    num_rows = len(df)\n",
        "#    for value in range(0, num_rows):\n",
        "#        # Convert a specific row to a dictionary (e.g., the first row)\n",
        "#        row_index = value\n",
        "#        row_dict = df.iloc[row_index].to_dict()\n",
        "#        if [df['UniProt_ID'] != UniProt_ID:\n",
        "#            # Convert dictionary to a list of dictionary items\n",
        "#            summaries.append(row_dict)\n",
        "#else:\n",
        "#    print(f\"File '{file_to_check}' does not exists.\")\n",
        "\n",
        "# Function to add a new entry to the summaries list\n",
        "def add_summary(UniProt_ID, No_of_all_available, No_of_PDB_available):\n",
        "    summary = {\n",
        "        \"UniProt_ID\": UniProt_ID,\n",
        "        \"No of predicted and exp-determined available\": No_of_all_available,\n",
        "        \"No of predicted available\": No_of_predicted_available,\n",
        "        \"No of exp-determined available\": No_of_exp_available,\n",
        "        \"No of PDB available\": No_of_PDB_available,\n",
        "        \"No of PDBs with 3A or better\": No_of_3A_and_better_available\n",
        "    }\n",
        "    summaries.append(summary)\n",
        "\n",
        "add_summary(UniProt_ID, No_of_all_available, No_of_PDB_available)\n",
        "\n",
        "# Convert the list of summaries to a pandas DataFrame\n",
        "df_summaries = pd.DataFrame(summaries)\n",
        "\n",
        "## Output the DataFrame to a CSV file\n",
        "#df_summaries.to_csv('struct_avail.csv', index=False)\n",
        "#print(\"DataFrame has been saved to struct_avail.csv file.\")\n",
        "#print()\n",
        "\n",
        "# Print the DataFrame\n",
        "#display(df_summaries)"
      ],
      "metadata": {
        "id": "MSrIzGLAhOOc",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "23608a61-fd51-4989-bb2b-1a46c4a71735",
        "cellView": "form"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "The following is a summary based on what has been filled-in so far.\n",
            "\n",
            "There are 34 structures (predicted and experimentally determined) available associated with UniProt ID P49768.\n",
            "There are 26 experimentally-determined structures associated with UniProt ID P49768.\n",
            "There are 26 experimentally-determined PDB structures associated with UniProt ID P49768.\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# @title **Assessing the available structures**\n",
        "\n",
        "UniProt_ID = 'P49768' # @param {type:\"string\"}\n",
        "# @markdown\n",
        "# @markdown ---\n",
        "# @markdown What are the PDB ids for the structural data  with the highest resolution that are available for this UniProt ID?\n",
        "Best_resolution = \"8kcs\" # @param {type:\"string\"}\n",
        "# @markdown HINT: Use **PDBe-KB** query\n",
        "# @markdown\n",
        "# @markdown ---\n",
        "# @markdown What are the PDB ids for the structural data with the most coverage available for this UniProt ID?\n",
        "Most_coverage = \"7d8x, 6lr4, 8kct, 7c9i, 7y5t, 6lqg, 8kcp, 6iyc, 6idf, 8x54, 8kcs, 8kco, 8k8e, 8im7, 8kcu, 8x53, 8oqy, 8x52, 8oqz, 5a63, 5fn4, 5fn5, 5fn3, 5fn2\" # @param {type:\"string\"}\n",
        "# @markdown HINT: Use **3D-Beacons** query\n",
        "# @markdown\n",
        "# @markdown ---\n",
        "# @markdown What are the PDB ids for the structural data without mutations available for this UniProt ID?\n",
        "Has_no_mutations = \"7d8x, 6lr4, 8kct, 7c9i, 7y5t, 6lqg, 8kcp, 8x54, 8kcs, 8kco, 8k8e, 8im7, 8kcu, 8x53, 8oqy, 8x52, 5a63, 5fn4, 5fn5, 2kr6\" # @param {type:\"string\"}\n",
        "# @markdown HINT: Use **PDBe-KB** query\n",
        "# @markdown\n",
        "# @markdown\n",
        "# @markdown ---\n",
        "\n",
        "if ',' in Best_resolution:\n",
        "  Best_resolution_list = Best_resolution.split(',')\n",
        "  Best_resolution_is = \"List\"\n",
        "else:\n",
        "  Best_resolution = Best_resolution\n",
        "  Best_resolution_is = \"Single\"\n",
        "\n",
        "if ',' in Most_coverage:\n",
        "  Most_coverage_list = Most_coverage.split(',')\n",
        "  Most_coverage_is = \"List\"\n",
        "else:\n",
        "  Most_coverage = Most_coverage\n",
        "  Most_coverage_is = \"Single\"\n",
        "\n",
        "if ',' in Has_no_mutations:\n",
        "  Has_no_mutations_list = Has_no_mutations.split(',')\n",
        "  Has_no_mutations_is = \"List\"\n",
        "else:\n",
        "  Has_no_mutations = Has_no_mutations\n",
        "  Has_no_mutations_is = \"Single\"\n",
        "\n",
        "\n",
        "# Print the assessment\n",
        "print(\"The following is a summary based on what has been filled-in so far.\")\n",
        "print()\n",
        "if Best_resolution_is == 'Single':\n",
        "  print(f\"There is a single structure with the best resolution associated with UniProt ID {UniProt_ID}.\")\n",
        "  print(f\"The PDB ID is {Best_resolution} for the best resolution structure.\")\n",
        "else:\n",
        "  print(f\"There are multiple structures with the best resolution associated with UniProt ID {UniProt_ID}.\")\n",
        "  print(f\"The PDB IDs of {Best_resolution_list} correspond to the best resolution structures.\")\n",
        "print()\n",
        "if Most_coverage_is == 'Single':\n",
        "  print(f\"There is a single structure with the most coverage of the UniProt ID {UniProt_ID}.\")\n",
        "  print(f\"The PDB ID is {Most_coverage} for the structure with most coverage of the UniProt sequence.\")\n",
        "else:\n",
        "  print(f\"There are multiple structures with the most coverage of the UniProt ID {UniProt_ID}.\")\n",
        "  print(f\"The PDB IDs of {Most_coverage_list} correspond to the structures with most coverage of the UniProt sequence.\")\n",
        "print()\n",
        "if Has_no_mutations_is == 'Single':\n",
        "  print(f\"There is a single structure with no mutations in sample sequence associated with UniProt ID {UniProt_ID}.\")\n",
        "  print(f\"The PDB ID is {Has_no_mutations} for a structure with no mutations.\")\n",
        "else:\n",
        "  print(f\"There are multiple structures with no mutations in sample sequence associated with UniProt ID {UniProt_ID}.\")\n",
        "  print(f\"The PDB IDs of {Has_no_mutations_list} correspond to structures with no mutations.\")\n",
        "print()\n",
        "\n",
        "# Initialize an empty list to store the assessments\n",
        "assessments = []\n",
        "\n",
        "## Define the file to check\n",
        "#file_to_check = \"/content/struct_assess.csv\"\n",
        "## Loop while the file exists\n",
        "#if os.path.exists(file_to_check):\n",
        "#    print(f\"File '{file_to_check}' exists.\")\n",
        "#    # Load the CSV file into a DataFrame\n",
        "#    df = pd.read_csv(file_to_check)\n",
        "#    # Get the number of rows\n",
        "#    num_rows = len(df)\n",
        "#    for value in range(0, num_rows):\n",
        "#        # Convert a specific row to a dictionary (e.g., the first row)\n",
        "#        row_index = value\n",
        "#        row_dict = df.iloc[row_index].to_dict()\n",
        "#        if [df['UniProt_ID'] != UniProt_ID:\n",
        "#            # Convert dictionary to a list of dictionary items\n",
        "#            assessments.append(row_dict)\n",
        "#else:\n",
        "#    print(f\"File '{file_to_check}' does not exists.\")\n",
        "\n",
        "# Function to add a new entry to the assessments list\n",
        "def add_assessment(UniProt_ID, Best_resolution, Most_coverage, Has_no_mutations):\n",
        "    def convert_to_list(input_str):\n",
        "        # Check if a comma is present in the string\n",
        "        if ',' in input_str:\n",
        "            # Split the string by commas and return the resulting list\n",
        "            return input_str.split(',')\n",
        "        else:\n",
        "            # Return the original string if no comma is present\n",
        "            return input_str\n",
        "    assessment = {\n",
        "        \"UniProt_ID\": UniProt_ID,\n",
        "        \"PDB id(s) with best resolution\": convert_to_list(Best_resolution),\n",
        "        \"PDB id(s) most coverage\": convert_to_list(Most_coverage),\n",
        "        \"PDB id(s) without any mutations\": convert_to_list(Has_no_mutations)\n",
        "    }\n",
        "    assessments.append(assessment)\n",
        "\n",
        "add_assessment(UniProt_ID, Best_resolution, Most_coverage, Has_no_mutations)\n",
        "\n",
        "# Convert the list of summaries to a pandas DataFrame\n",
        "df_assessments = pd.DataFrame(assessments)\n",
        "\n",
        "## Output the DataFrame to a CSV file\n",
        "#df_summaries.to_csv('struct_assess.csv', index=False)\n",
        "#print(\"DataFrame has been saved to struct_assess.csv file.\")\n",
        "#print()\n",
        "\n",
        "# Print the DataFrame\n",
        "#display(df_assessments)"
      ],
      "metadata": {
        "id": "HEghuXE4SbPe",
        "cellView": "form",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "9b75c224-39e5-4c05-9b4b-8bb3df84b231"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "The following is a summary based on what has been filled-in so far.\n",
            "\n",
            "There is a single structure with the best resolution associated with UniProt ID P49768.\n",
            "The PDB ID is 8kcs for the best resolution structure.\n",
            "\n",
            "There are multiple structures with the most coverage of the UniProt ID P49768.\n",
            "The PDB IDs of ['7d8x', ' 6lr4', ' 8kct', ' 7c9i', ' 7y5t', ' 6lqg', ' 8kcp', ' 6iyc', ' 6idf', ' 8x54', ' 8kcs', ' 8kco', ' 8k8e', ' 8im7', ' 8kcu', ' 8x53', ' 8oqy', ' 8x52', ' 8oqz', ' 5a63', ' 5fn4', ' 5fn5', ' 5fn3', ' 5fn2'] correspond to the structures with most coverage of the UniProt sequence.\n",
            "\n",
            "There are multiple structures with no mutations in sample sequence associated with UniProt ID P49768.\n",
            "The PDB IDs of ['7d8x', ' 6lr4', ' 8kct', ' 7c9i', ' 7y5t', ' 6lqg', ' 8kcp', ' 8x54', ' 8kcs', ' 8kco', ' 8k8e', ' 8im7', ' 8kcu', ' 8x53', ' 8oqy', ' 8x52', ' 5a63', ' 5fn4', ' 5fn5', ' 2kr6'] correspond to structures with no mutations.\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# @title **Structure summary**\n",
        "\n",
        "UniProt_ID = 'P49768' # @param {type:\"string\"}\n",
        "# @markdown\n",
        "# @markdown ---\n",
        "##\n",
        "\n",
        "filter_value = UniProt_ID\n",
        "\n",
        "# Filter DataFrames based on the 'UniProt_ID' column value\n",
        "filtered_df_summaries = df_summaries[df_summaries['UniProt_ID'] == filter_value]\n",
        "filtered_df_assessments = df_assessments[df_assessments['UniProt_ID'] == filter_value]\n",
        "\n",
        "# Merge filtered DataFrames based on shared value in 'UniProt_ID' column\n",
        "df_structures = pd.merge(filtered_df_summaries, filtered_df_assessments, on='UniProt_ID')\n",
        "\n",
        "# Transpose the resulting DataFrame\n",
        "df_structures_transposed = df_structures.T\n",
        "\n",
        "\n",
        "# Left-align the index\n",
        "df_formatted = df_structures_transposed.style.set_table_styles(\n",
        "    [{'selector': 'th.row_heading', 'props': [('text-align', 'left')]}],\n",
        ")\n",
        "\n",
        "print(\"Table 1: General Structure Availability\")\n",
        "display(df_formatted)"
      ],
      "metadata": {
        "id": "Huilt06PjZRG",
        "cellView": "form",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 350
        },
        "outputId": "325ba470-89b4-4986-b1e8-993436acdfce"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Table 1: General Structure Availability\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<pandas.io.formats.style.Styler at 0x79f2ad38e710>"
            ],
            "text/html": [
              "<style type=\"text/css\">\n",
              "#T_240cf th.row_heading {\n",
              "  text-align: left;\n",
              "}\n",
              "</style>\n",
              "<table id=\"T_240cf\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr>\n",
              "      <th class=\"blank level0\" >&nbsp;</th>\n",
              "      <th id=\"T_240cf_level0_col0\" class=\"col_heading level0 col0\" >0</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th id=\"T_240cf_level0_row0\" class=\"row_heading level0 row0\" >UniProt_ID</th>\n",
              "      <td id=\"T_240cf_row0_col0\" class=\"data row0 col0\" >P49768</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th id=\"T_240cf_level0_row1\" class=\"row_heading level0 row1\" >No of predicted and exp-determined available</th>\n",
              "      <td id=\"T_240cf_row1_col0\" class=\"data row1 col0\" >34</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th id=\"T_240cf_level0_row2\" class=\"row_heading level0 row2\" >No of predicted available</th>\n",
              "      <td id=\"T_240cf_row2_col0\" class=\"data row2 col0\" >8</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th id=\"T_240cf_level0_row3\" class=\"row_heading level0 row3\" >No of exp-determined available</th>\n",
              "      <td id=\"T_240cf_row3_col0\" class=\"data row3 col0\" >26</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th id=\"T_240cf_level0_row4\" class=\"row_heading level0 row4\" >No of PDB available</th>\n",
              "      <td id=\"T_240cf_row4_col0\" class=\"data row4 col0\" >26</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th id=\"T_240cf_level0_row5\" class=\"row_heading level0 row5\" >No of PDBs with 3A or better</th>\n",
              "      <td id=\"T_240cf_row5_col0\" class=\"data row5 col0\" >14</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th id=\"T_240cf_level0_row6\" class=\"row_heading level0 row6\" >PDB id(s) with best resolution</th>\n",
              "      <td id=\"T_240cf_row6_col0\" class=\"data row6 col0\" >8kcs</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th id=\"T_240cf_level0_row7\" class=\"row_heading level0 row7\" >PDB id(s) most coverage</th>\n",
              "      <td id=\"T_240cf_row7_col0\" class=\"data row7 col0\" >['7d8x', ' 6lr4', ' 8kct', ' 7c9i', ' 7y5t', ' 6lqg', ' 8kcp', ' 6iyc', ' 6idf', ' 8x54', ' 8kcs', ' 8kco', ' 8k8e', ' 8im7', ' 8kcu', ' 8x53', ' 8oqy', ' 8x52', ' 8oqz', ' 5a63', ' 5fn4', ' 5fn5', ' 5fn3', ' 5fn2']</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th id=\"T_240cf_level0_row8\" class=\"row_heading level0 row8\" >PDB id(s) without any mutations</th>\n",
              "      <td id=\"T_240cf_row8_col0\" class=\"data row8 col0\" >['7d8x', ' 6lr4', ' 8kct', ' 7c9i', ' 7y5t', ' 6lqg', ' 8kcp', ' 8x54', ' 8kcs', ' 8kco', ' 8k8e', ' 8im7', ' 8kcu', ' 8x53', ' 8oqy', ' 8x52', ' 5a63', ' 5fn4', ' 5fn5', ' 2kr6']</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n"
            ]
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Variants"
      ],
      "metadata": {
        "id": "V-KqJR8JFq1r"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# @title **Input variants**\n",
        "# @markdown\n",
        "# @markdown Put variants being considered in textbox below.\n",
        "# @markdown\n",
        "# @markdown HINT: Use format where **(1st)** amino acid change, and **(2nd)** position in sequence.\n",
        "\n",
        "# @markdown ---\n",
        "##\n",
        "\n",
        "text_input = widgets.Text(\n",
        "    value='',\n",
        "    placeholder='e.g. 6  K/M',\n",
        "    description='Input:',\n",
        "    disabled=False\n",
        ")\n",
        "\n",
        "display(text_input)\n",
        "\n",
        "def on_text_change(change):\n",
        "    print(f'You entered: {change[\"new\"]}')\n",
        "\n",
        "text_input.observe(on_text_change, names='value')"
      ],
      "metadata": {
        "id": "Jk5q-8nE8nBp",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 67,
          "referenced_widgets": [
            "0a50d17db302436086d575423938a570",
            "85210f49294149fabd66ec9f27d4076e",
            "f58043a3bbe14371a997d86d5d6f1bf8"
          ]
        },
        "cellView": "form",
        "outputId": "271cefaa-0931-443a-a516-5b3c34245e89"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "Text(value='', description='Input:', placeholder='e.g. 6  K/M')"
            ],
            "application/vnd.jupyter.widget-view+json": {
              "version_major": 2,
              "version_minor": 0,
              "model_id": "0a50d17db302436086d575423938a570"
            }
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "You entered: 435 L/F\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# @title **Press shift + enter to convert input to table**\n",
        "\n",
        "input = text_input.value\n",
        "print(f\"Input is: {input}\")\n",
        "# Split the string into parts, assuming space separates variants\n",
        "parts = input.split()\n",
        "\n",
        "# Create a list of tuples (residue_number, amino_acid_change)\n",
        "# Iterate through the parts in steps of 2, taking the residue number (first part) and the amino acid change (second part)\n",
        "data = [(int(parts[i]), parts[i+1]) for i in range(0, len(parts), 2)]\n",
        "\n",
        "# Create DataFrame\n",
        "df_variants = pd.DataFrame(data, columns=[\"residue_number\", \"amino_acid_change\"])\n",
        "\n",
        "# Split the 'amino_acid_change' column into two new columns\n",
        "df_variants[[\"original_aa\", \"mutated_aa\"]] = df_variants[\"amino_acid_change\"].str.split(\"/\", expand=True)\n",
        "\n",
        "# Optional: Reorder or drop columns if needed\n",
        "df_variants = df_variants[[\"original_aa\", \"mutated_aa\", \"residue_number\"]]\n",
        "\n",
        "# Dictionary for one-letter to three-letter amino acid codes\n",
        "aa_dict = {\n",
        "    \"A\": \"ALA\", \"R\": \"ARG\", \"N\": \"ASN\", \"D\": \"ASP\", \"C\": \"CYS\",\n",
        "    \"Q\": \"GLN\", \"E\": \"GLU\", \"G\": \"GLY\", \"H\": \"HIS\", \"I\": \"ILE\",\n",
        "    \"L\": \"LEU\", \"K\": \"LYS\", \"M\": \"MET\", \"F\": \"PHE\", \"P\": \"PRO\",\n",
        "    \"S\": \"SER\", \"T\": \"THR\", \"W\": \"TRP\", \"Y\": \"TYR\", \"V\": \"VAL\"\n",
        "}\n",
        "\n",
        "# Map the one-letter codes to three-letter codes\n",
        "df_variants[\"original_aa\"] = df_variants[\"original_aa\"].map(aa_dict)\n",
        "df_variants[\"mutated_aa\"] = df_variants[\"mutated_aa\"].map(aa_dict)\n",
        "\n",
        "# Add a column with row numbers starting from 1\n",
        "df_variants.insert(0, \"no.\", range(1, len(df_variants) + 1))\n",
        "\n",
        "display(df_variants)"
      ],
      "metadata": {
        "id": "RPekMmeL1ja9",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 107
        },
        "cellView": "form",
        "outputId": "f6f3b8d1-81ff-406c-e162-97d85ec9407b"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Input is: 435 L/F\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "   no. original_aa mutated_aa  residue_number\n",
              "0    1         LEU        PHE             435"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-6dcf1cfc-681b-43cd-9cdd-68ca7cb6f04d\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>no.</th>\n",
              "      <th>original_aa</th>\n",
              "      <th>mutated_aa</th>\n",
              "      <th>residue_number</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>1</td>\n",
              "      <td>LEU</td>\n",
              "      <td>PHE</td>\n",
              "      <td>435</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-6dcf1cfc-681b-43cd-9cdd-68ca7cb6f04d')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-6dcf1cfc-681b-43cd-9cdd-68ca7cb6f04d button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-6dcf1cfc-681b-43cd-9cdd-68ca7cb6f04d');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "  <div id=\"id_4faafa0a-2a06-4d9a-affe-13efa4a3f42f\">\n",
              "    <style>\n",
              "      .colab-df-generate {\n",
              "        background-color: #E8F0FE;\n",
              "        border: none;\n",
              "        border-radius: 50%;\n",
              "        cursor: pointer;\n",
              "        display: none;\n",
              "        fill: #1967D2;\n",
              "        height: 32px;\n",
              "        padding: 0 0 0 0;\n",
              "        width: 32px;\n",
              "      }\n",
              "\n",
              "      .colab-df-generate:hover {\n",
              "        background-color: #E2EBFA;\n",
              "        box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "        fill: #174EA6;\n",
              "      }\n",
              "\n",
              "      [theme=dark] .colab-df-generate {\n",
              "        background-color: #3B4455;\n",
              "        fill: #D2E3FC;\n",
              "      }\n",
              "\n",
              "      [theme=dark] .colab-df-generate:hover {\n",
              "        background-color: #434B5C;\n",
              "        box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "        filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "        fill: #FFFFFF;\n",
              "      }\n",
              "    </style>\n",
              "    <button class=\"colab-df-generate\" onclick=\"generateWithVariable('df_variants')\"\n",
              "            title=\"Generate code using this dataframe.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M7,19H8.4L18.45,9,17,7.55,7,17.6ZM5,21V16.75L18.45,3.32a2,2,0,0,1,2.83,0l1.4,1.43a1.91,1.91,0,0,1,.58,1.4,1.91,1.91,0,0,1-.58,1.4L9.25,21ZM18.45,9,17,7.55Zm-12,3A5.31,5.31,0,0,0,4.9,8.1,5.31,5.31,0,0,0,1,6.5,5.31,5.31,0,0,0,4.9,4.9,5.31,5.31,0,0,0,6.5,1,5.31,5.31,0,0,0,8.1,4.9,5.31,5.31,0,0,0,12,6.5,5.46,5.46,0,0,0,6.5,12Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "    <script>\n",
              "      (() => {\n",
              "      const buttonEl =\n",
              "        document.querySelector('#id_4faafa0a-2a06-4d9a-affe-13efa4a3f42f button.colab-df-generate');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      buttonEl.onclick = () => {\n",
              "        google.colab.notebook.generateWithVariable('df_variants');\n",
              "      }\n",
              "      })();\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "variable_name": "df_variants",
              "summary": "{\n  \"name\": \"df_variants\",\n  \"rows\": 1,\n  \"fields\": [\n    {\n      \"column\": \"no.\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": null,\n        \"min\": 1,\n        \"max\": 1,\n        \"num_unique_values\": 1,\n        \"samples\": [\n          1\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"original_aa\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 1,\n        \"samples\": [\n          \"LEU\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"mutated_aa\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 1,\n        \"samples\": [\n          \"PHE\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"residue_number\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": null,\n        \"min\": 435,\n        \"max\": 435,\n        \"num_unique_values\": 1,\n        \"samples\": [\n          435\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}"
            }
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## BLOSUM62 Matrix\n",
        "<img src=\"https://upload.wikimedia.org/wikipedia/commons/f/f5/Blosum62-dayhoff-ordering.svg\" height=\"500\" align=\"right\">\n",
        "\n",
        "The BLOSUM62 (BLOcks SUbstitution Matrix) is a substitution matrix that can be used in alignment of protein sequences (*e.g.* [BLASTp](https://blast.ncbi.nlm.nih.gov/Blast.cgi?PROGRAM=blastp&PAGE_TYPE=BlastSearch&LINK_LOC=blasthome) and [UniProt's BLAST](https://www.uniprot.org/blast)). It was generated from analysis of many protein sequences.\n",
        "\n",
        "<br>\n",
        "\n",
        "Positive values indicate frequent exchanges between amino acid pairs during evolution, negative values indicate amino acid pairs that rarely replace each other. The colouring in the image of BLOSUM62 is from Margaret Dayhoff's amino acid classification. Together these highlight that amino acids can be clustered in terms of more and less similar.\n",
        "\n",
        "<br>\n",
        "\n",
        "For an overview of the development of BLOSUM, click [here](https://en.wikipedia.org/wiki/BLOSUM)<br>\n",
        "\n",
        "<br>\n",
        "\n",
        "\n"
      ],
      "metadata": {
        "id": "6PJIwLTRUiSt"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "## **Table of Margaret Dayhoff's encoding of amino acids**\n",
        "\n",
        "|                     Amino acids                      | 1-letter code  |      3-letter code       |        Property        | Dayhoff  |\n",
        "|:----------------------------------------------------:|:--------------:|:------------------------:|:----------------------:|:--------:|\n",
        "| Cysteine                                             | C              | Cys                      | Sulphur polymerization  | a        |\n",
        "| Glycine, Serine, Threonine, Alanine, Proline         | G, S, T, A, P  | Gly, Ser, Thr, Ala, Pro  | Small                  | b        |\n",
        "| Aspartic acid, Glutamic acid, Asparagine, Glutamine  | D, E, N, Q     | Asp, Glu, Asn, Gln       | Acid and amide         | c        |\n",
        "| Arginine, Histidine, Lysine                          | R, H, K        | Arg, His, Lys            | Basic                  | d        |\n",
        "| Leucine, Valine, Methionine, Isoleucine              | L, V, M, I     | Leu, Val, Met, Ile       | Hydrophobic            | e        |\n",
        "| Tyrosine, Phenylalanine, Tryptophan                  | Y, F, W        | Tyr, Phe, Trp            | Aromatic               | f        |"
      ],
      "metadata": {
        "id": "0H9dVT5JB-_E"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# @title **Press shift + enter to add Dayhoff information to table**\n",
        "\n",
        "# Define Dayhoff groups\n",
        "dayhoff_groups = {\n",
        "    'GLY': '1', 'ALA': '1', 'SER': '1', 'THR': '1', 'PRO': '1',\n",
        "    'ASP': '2', 'GLU': '2',\n",
        "    'ASN': '3', 'GLN': '3',\n",
        "    'ARG': '4', 'LYS': '4', 'HIS': '4',\n",
        "    'MET': '5', 'ILE': '5', 'LEU': '5', 'VAL': '5', 'CYS': '5',\n",
        "    'PHE': '6', 'TYR': '6', 'TRP': '6'\n",
        "}\n",
        "\n",
        "\n",
        "# Map amino acids to Dayhoff groups\n",
        "df_variants['Dayhoff_for_original_aa'] = df_variants['original_aa'].map(dayhoff_groups)\n",
        "df_variants['Dayhoff_for_mutated_aa'] = df_variants['mutated_aa'].map(dayhoff_groups)\n",
        "\n",
        "# Compare groups\n",
        "df_variants['SameGroup'] = df_variants['Dayhoff_for_original_aa'] == df_variants['Dayhoff_for_mutated_aa']\n",
        "df_variants['SameGroup'] = df_variants['SameGroup'].map({True: 'Yes', False: 'No'})\n",
        "\n",
        "display(df_variants)"
      ],
      "metadata": {
        "id": "o0KrpD4FqIry",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 89
        },
        "cellView": "form",
        "outputId": "590f8524-8f3f-496c-caf2-9bab8ca20e31"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "   no. original_aa mutated_aa  residue_number Dayhoff_for_original_aa  \\\n",
              "0    1         LEU        PHE             435                       5   \n",
              "\n",
              "  Dayhoff_for_mutated_aa SameGroup  \n",
              "0                      6        No  "
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-e5d2a27d-9e79-4c1e-8dfe-8a2724e5601c\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>no.</th>\n",
              "      <th>original_aa</th>\n",
              "      <th>mutated_aa</th>\n",
              "      <th>residue_number</th>\n",
              "      <th>Dayhoff_for_original_aa</th>\n",
              "      <th>Dayhoff_for_mutated_aa</th>\n",
              "      <th>SameGroup</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>1</td>\n",
              "      <td>LEU</td>\n",
              "      <td>PHE</td>\n",
              "      <td>435</td>\n",
              "      <td>5</td>\n",
              "      <td>6</td>\n",
              "      <td>No</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-e5d2a27d-9e79-4c1e-8dfe-8a2724e5601c')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-e5d2a27d-9e79-4c1e-8dfe-8a2724e5601c button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-e5d2a27d-9e79-4c1e-8dfe-8a2724e5601c');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "  <div id=\"id_11f84791-2539-42be-8a49-6c36297fe8ae\">\n",
              "    <style>\n",
              "      .colab-df-generate {\n",
              "        background-color: #E8F0FE;\n",
              "        border: none;\n",
              "        border-radius: 50%;\n",
              "        cursor: pointer;\n",
              "        display: none;\n",
              "        fill: #1967D2;\n",
              "        height: 32px;\n",
              "        padding: 0 0 0 0;\n",
              "        width: 32px;\n",
              "      }\n",
              "\n",
              "      .colab-df-generate:hover {\n",
              "        background-color: #E2EBFA;\n",
              "        box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "        fill: #174EA6;\n",
              "      }\n",
              "\n",
              "      [theme=dark] .colab-df-generate {\n",
              "        background-color: #3B4455;\n",
              "        fill: #D2E3FC;\n",
              "      }\n",
              "\n",
              "      [theme=dark] .colab-df-generate:hover {\n",
              "        background-color: #434B5C;\n",
              "        box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "        filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "        fill: #FFFFFF;\n",
              "      }\n",
              "    </style>\n",
              "    <button class=\"colab-df-generate\" onclick=\"generateWithVariable('df_variants')\"\n",
              "            title=\"Generate code using this dataframe.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M7,19H8.4L18.45,9,17,7.55,7,17.6ZM5,21V16.75L18.45,3.32a2,2,0,0,1,2.83,0l1.4,1.43a1.91,1.91,0,0,1,.58,1.4,1.91,1.91,0,0,1-.58,1.4L9.25,21ZM18.45,9,17,7.55Zm-12,3A5.31,5.31,0,0,0,4.9,8.1,5.31,5.31,0,0,0,1,6.5,5.31,5.31,0,0,0,4.9,4.9,5.31,5.31,0,0,0,6.5,1,5.31,5.31,0,0,0,8.1,4.9,5.31,5.31,0,0,0,12,6.5,5.46,5.46,0,0,0,6.5,12Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "    <script>\n",
              "      (() => {\n",
              "      const buttonEl =\n",
              "        document.querySelector('#id_11f84791-2539-42be-8a49-6c36297fe8ae button.colab-df-generate');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      buttonEl.onclick = () => {\n",
              "        google.colab.notebook.generateWithVariable('df_variants');\n",
              "      }\n",
              "      })();\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "variable_name": "df_variants",
              "summary": "{\n  \"name\": \"df_variants\",\n  \"rows\": 1,\n  \"fields\": [\n    {\n      \"column\": \"no.\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": null,\n        \"min\": 1,\n        \"max\": 1,\n        \"num_unique_values\": 1,\n        \"samples\": [\n          1\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"original_aa\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 1,\n        \"samples\": [\n          \"LEU\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"mutated_aa\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 1,\n        \"samples\": [\n          \"PHE\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"residue_number\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": null,\n        \"min\": 435,\n        \"max\": 435,\n        \"num_unique_values\": 1,\n        \"samples\": [\n          435\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Dayhoff_for_original_aa\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 1,\n        \"samples\": [\n          \"5\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Dayhoff_for_mutated_aa\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 1,\n        \"samples\": [\n          \"6\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"SameGroup\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 1,\n        \"samples\": [\n          \"No\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}"
            }
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# @title Questions\n",
        "# @markdown\n",
        "# @markdown\n",
        "# @markdown ---\n",
        "# @markdown What is the BLOSUM62 value for the canonical to variant change?\n",
        "BLOSUM62_value = 0 # @param {type:\"number\"}\n",
        "# @markdown ---\n",
        "# @markdown If the canonical and variant amino acids are in the same Dayhoff group, which one?\n",
        "Dayhoff_group = 'Not_same_group' # @param [\"Not_same_group\", \"Sulfur_polymerization\",\"Small\", \"Acid_and_amide\", \"Basic\", \"Hydrophobic\", \"Aromatic\"]\n",
        "# @markdown ---"
      ],
      "metadata": {
        "id": "0PdWwk7J_Zsw",
        "cellView": "form"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Amino acid details\n",
        "\n",
        "---\n"
      ],
      "metadata": {
        "id": "HRQ_wRC8560m"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "Not all genetic variants resulted in an amino acid change, but here we will be focusing on the genetic variants that cause an amino acid change (missense variants).\n",
        "\n",
        "We will be contrasting the amino acid present in the genetic variant, with the canonical amino acid. A canonical amino acid is the one most commonly observed at a particular position in the protein sequence."
      ],
      "metadata": {
        "id": "fyOyQ6bYD6U-"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# @title **Amino acid change input if not using table**\n",
        "# @markdown\n",
        "# @markdown Not all genetic variants resulted in an amino acid change, but here we will be focusing on the genetic variants that cause an amino acid change (missense variants).\n",
        "# @markdown\n",
        "# @markdown\n",
        "# @markdown We will be contrasting the amino acid present in the genetic variant, with the canonical amino acid.\n",
        "# @markdown A canonical amino acid is the one most commonly observed at a particular position in the protein sequence.\n",
        "# @markdown\n",
        "# @markdown **Pressing run / shift-enter** on this section captures the identity of the amino acids and sets up the code for the subsequent visuallizations on amino acid type.\n",
        "# @markdown\n",
        "# @markdown ---\n",
        "# @markdown For the variant of interest, <br>what was the canonical amino acid?\n",
        "Canonical_Amino_Acid = 'Select the amino acid' # @param ['Select the amino acid', 'GLY (G)', 'ALA (A)', 'VAL (V)', 'LEU (L)', 'ILE (I)', 'THR (T)', 'SER (S)', 'MET (M)', 'CYS (C)', 'PRO (P)', 'PHE (F)', 'TYR (Y)', 'TRP (W)', 'HIS (H)', 'LYS (K)', 'ARG (R)', 'ASP (D)', 'GLU (E)', 'ASN (N)', 'GLN (Q)']\n",
        "# @markdown ---\n",
        "# @markdown For the variant of interest, <br>what was its position in the amino acid sequence (in UniProt numbering)?\n",
        "Variant_Position = 0 # @param {type:\"number\"}\n",
        "# @markdown ---\n",
        "# @markdown For the variant of interest, <br>what was the new amino acid resulting from the genetic variant?\n",
        "New_Amino_Acid = 'Select the amino acid' # @param ['Select the amino acid', 'GLY (G)', 'ALA (A)', 'VAL (V)', 'LEU (L)', 'ILE (I)', 'THR (T)', 'SER (S)', 'MET (M)', 'CYS (C)', 'PRO (P)', 'PHE (F)', 'TYR (Y)', 'TRP (W)', 'HIS (H)', 'LYS (K)', 'ARG (R)', 'ASP (D)', 'GLU (E)', 'ASN (N)', 'GLN (Q)']\n",
        "# @markdown ---\n",
        "\n",
        "####### COLOURS\n",
        "#defining colour variables based on amino acids identity\n",
        "\n",
        "if Canonical_Amino_Acid == 'CYS (C)' or Canonical_Amino_Acid == 'PRO (P)' or Canonical_Amino_Acid == 'GLY (G)':\n",
        "    colour1 = '#cccc00'\n",
        "    colourword1 = 'yellow'\n",
        "elif Canonical_Amino_Acid == 'SER (S)' or Canonical_Amino_Acid == 'THR (T)' or Canonical_Amino_Acid == 'ASN (N)' or Canonical_Amino_Acid == 'GLN (Q)':\n",
        "    colour1 = '#cc99ff'\n",
        "    colourword1 = 'pink-purple'\n",
        "elif Canonical_Amino_Acid == 'ARG (R)' or Canonical_Amino_Acid == 'HIS (H)' or Canonical_Amino_Acid == 'LYS (K)' or Canonical_Amino_Acid == 'ASP (D)' or Canonical_Amino_Acid == 'GLU (E)':\n",
        "    colour1 = '#9999ff'\n",
        "    colourword1 = 'blue-purple'\n",
        "elif Canonical_Amino_Acid == 'ALA (A)' or Canonical_Amino_Acid == 'VAL (V)' or Canonical_Amino_Acid == 'ILE (I)' or Canonical_Amino_Acid == 'LEU (L)':\n",
        "    colour1 = '#99ff66'\n",
        "    colourword1 = 'green'\n",
        "elif Canonical_Amino_Acid == 'MET (M)' or Canonical_Amino_Acid == 'PHE (F)' or Canonical_Amino_Acid == 'TYR (Y)' or Canonical_Amino_Acid == 'TRP (W)':\n",
        "    colour1 = '#99ff66'\n",
        "    colourword1 = 'green'\n",
        "else:\n",
        "    colour1 = '#b3b3b3'\n",
        "\n",
        "if New_Amino_Acid == 'CYS (C)' or New_Amino_Acid == 'PRO (P)' or New_Amino_Acid == 'GLY (G)':\n",
        "    colour2 = '#cccc00'\n",
        "    colourword2 = 'yellow'\n",
        "elif New_Amino_Acid == 'SER (S)' or New_Amino_Acid == 'THR (T)' or New_Amino_Acid == 'ASN (N)' or New_Amino_Acid == 'GLN (Q)':\n",
        "    colour2 = '#cc99ff'\n",
        "    colourword2 = 'pink-purple'\n",
        "elif New_Amino_Acid == 'ARG (R)' or New_Amino_Acid == 'HIS (H)' or New_Amino_Acid == 'LYS (K)' or New_Amino_Acid == 'ASP (D)' or New_Amino_Acid == 'GLU (E)':\n",
        "    colour2 = '#9999ff'\n",
        "    colourword2 = 'blue-purple'\n",
        "elif New_Amino_Acid == 'ALA (A)' or New_Amino_Acid == 'VAL (V)' or New_Amino_Acid == 'ILE (I)' or New_Amino_Acid == 'LEU (L)':\n",
        "    colour2 = '#99ff66'\n",
        "    colourword2 = 'green'\n",
        "elif New_Amino_Acid == 'MET (M)' or New_Amino_Acid == 'PHE (F)' or New_Amino_Acid == 'TYR (Y)' or New_Amino_Acid == 'TRP (W)':\n",
        "    colour2 = '#99ff66'\n",
        "    colourword2 = 'green'\n",
        "else:\n",
        "    colour2 = '#b3b3b3'\n",
        "\n",
        "if Canonical_Amino_Acid == 'CYS (C)' or Canonical_Amino_Acid == 'ASP (D)' or Canonical_Amino_Acid == 'GLU (E)' or Canonical_Amino_Acid == 'THR (T)':\n",
        "    colour3 = 'red'\n",
        "    colourword3 = 'red'\n",
        "elif Canonical_Amino_Acid == 'SER (S)' or Canonical_Amino_Acid == 'TYR (Y)':\n",
        "    colour3 = '#ff6666'\n",
        "    colourword3 = 'light red'\n",
        "elif Canonical_Amino_Acid == 'ARG (R)' or Canonical_Amino_Acid == 'LYS (K)':\n",
        "    colour3 = 'blue'\n",
        "    colourword3 = 'blue'\n",
        "elif  Canonical_Amino_Acid == 'HIS (H)' or Canonical_Amino_Acid == 'TRP (W)' or Canonical_Amino_Acid == 'ASN (N)' or Canonical_Amino_Acid == 'GLN (Q)':\n",
        "    colour3 = '#6699ff'\n",
        "    colourword3 = 'light blue'\n",
        "elif Canonical_Amino_Acid == 'ALA (A)' or Canonical_Amino_Acid == 'VAL (V)' or Canonical_Amino_Acid == 'ILE (I)' or Canonical_Amino_Acid == 'LEU (L)':\n",
        "    colour3 = '#b3b3b3'\n",
        "    colourword3 = 'grey'\n",
        "elif Canonical_Amino_Acid == 'MET (M)' or Canonical_Amino_Acid == 'PHE (F)' or Canonical_Amino_Acid == 'PRO (P)':\n",
        "    colour3 = '#b3b3b3'\n",
        "    colourword3 = 'grey'\n",
        "else:\n",
        "    colour3 = '#b3b3b3'\n",
        "    colourword3 = 'grey'\n",
        "\n",
        "if New_Amino_Acid == 'CYS (C)' or New_Amino_Acid == 'ASP (D)' or New_Amino_Acid == 'GLU (E)' or New_Amino_Acid == 'THR (T)':\n",
        "    colour4 = 'red'\n",
        "    colourword4 = 'red'\n",
        "elif New_Amino_Acid == 'SER (S)' or New_Amino_Acid == 'TYR (Y)':\n",
        "    colour4 = '#ff6666'\n",
        "    colourword4 = 'light red'\n",
        "elif New_Amino_Acid == 'ARG (R)' or New_Amino_Acid == 'LYS (K)':\n",
        "    colour4 = 'blue'\n",
        "    colourword4 = 'blue'\n",
        "elif  New_Amino_Acid == 'HIS (H)' or New_Amino_Acid == 'TRP (W)' or New_Amino_Acid == 'ASN (N)' or New_Amino_Acid == 'GLN (Q)':\n",
        "    colour4 = '#6699ff'\n",
        "    colourword4 = 'light blue'\n",
        "elif New_Amino_Acid == 'ALA (A)' or New_Amino_Acid == 'VAL (V)' or New_Amino_Acid == 'ILE (I)' or New_Amino_Acid == 'LEU (L)':\n",
        "    colour4 = '#b3b3b3'\n",
        "    colourword4 = 'grey'\n",
        "elif New_Amino_Acid == 'MET (M)' or New_Amino_Acid == 'PHE (F)' or New_Amino_Acid == 'PRO (P)':\n",
        "    colour4 = '#b3b3b3'\n",
        "    colourword4 = 'grey'\n",
        "else:\n",
        "    colour4 = '#b3b3b3'\n",
        "    colourword4 = 'grey'"
      ],
      "metadata": {
        "cellView": "form",
        "id": "ilttYMwfDUnQ"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# @title **View the Variants table**\n",
        "\n",
        "display(df_variants)"
      ],
      "metadata": {
        "id": "9nZqoARyEDM0",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 89
        },
        "cellView": "form",
        "outputId": "5df8e015-b7c9-483e-fb20-c87aeb6dac85"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "   no. original_aa mutated_aa  residue_number Dayhoff_for_original_aa  \\\n",
              "0    1         LEU        PHE             435                       5   \n",
              "\n",
              "  Dayhoff_for_mutated_aa SameGroup  \n",
              "0                      6        No  "
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-37311fe5-cd3a-4d08-b41e-a9f86c203ada\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>no.</th>\n",
              "      <th>original_aa</th>\n",
              "      <th>mutated_aa</th>\n",
              "      <th>residue_number</th>\n",
              "      <th>Dayhoff_for_original_aa</th>\n",
              "      <th>Dayhoff_for_mutated_aa</th>\n",
              "      <th>SameGroup</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>1</td>\n",
              "      <td>LEU</td>\n",
              "      <td>PHE</td>\n",
              "      <td>435</td>\n",
              "      <td>5</td>\n",
              "      <td>6</td>\n",
              "      <td>No</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-37311fe5-cd3a-4d08-b41e-a9f86c203ada')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-37311fe5-cd3a-4d08-b41e-a9f86c203ada button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-37311fe5-cd3a-4d08-b41e-a9f86c203ada');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "  <div id=\"id_534d3a08-f8dd-4458-9a25-89adc515b269\">\n",
              "    <style>\n",
              "      .colab-df-generate {\n",
              "        background-color: #E8F0FE;\n",
              "        border: none;\n",
              "        border-radius: 50%;\n",
              "        cursor: pointer;\n",
              "        display: none;\n",
              "        fill: #1967D2;\n",
              "        height: 32px;\n",
              "        padding: 0 0 0 0;\n",
              "        width: 32px;\n",
              "      }\n",
              "\n",
              "      .colab-df-generate:hover {\n",
              "        background-color: #E2EBFA;\n",
              "        box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "        fill: #174EA6;\n",
              "      }\n",
              "\n",
              "      [theme=dark] .colab-df-generate {\n",
              "        background-color: #3B4455;\n",
              "        fill: #D2E3FC;\n",
              "      }\n",
              "\n",
              "      [theme=dark] .colab-df-generate:hover {\n",
              "        background-color: #434B5C;\n",
              "        box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "        filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "        fill: #FFFFFF;\n",
              "      }\n",
              "    </style>\n",
              "    <button class=\"colab-df-generate\" onclick=\"generateWithVariable('df_variants')\"\n",
              "            title=\"Generate code using this dataframe.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M7,19H8.4L18.45,9,17,7.55,7,17.6ZM5,21V16.75L18.45,3.32a2,2,0,0,1,2.83,0l1.4,1.43a1.91,1.91,0,0,1,.58,1.4,1.91,1.91,0,0,1-.58,1.4L9.25,21ZM18.45,9,17,7.55Zm-12,3A5.31,5.31,0,0,0,4.9,8.1,5.31,5.31,0,0,0,1,6.5,5.31,5.31,0,0,0,4.9,4.9,5.31,5.31,0,0,0,6.5,1,5.31,5.31,0,0,0,8.1,4.9,5.31,5.31,0,0,0,12,6.5,5.46,5.46,0,0,0,6.5,12Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "    <script>\n",
              "      (() => {\n",
              "      const buttonEl =\n",
              "        document.querySelector('#id_534d3a08-f8dd-4458-9a25-89adc515b269 button.colab-df-generate');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      buttonEl.onclick = () => {\n",
              "        google.colab.notebook.generateWithVariable('df_variants');\n",
              "      }\n",
              "      })();\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "variable_name": "df_variants",
              "summary": "{\n  \"name\": \"df_variants\",\n  \"rows\": 1,\n  \"fields\": [\n    {\n      \"column\": \"no.\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": null,\n        \"min\": 1,\n        \"max\": 1,\n        \"num_unique_values\": 1,\n        \"samples\": [\n          1\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"original_aa\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 1,\n        \"samples\": [\n          \"LEU\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"mutated_aa\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 1,\n        \"samples\": [\n          \"PHE\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"residue_number\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": null,\n        \"min\": 435,\n        \"max\": 435,\n        \"num_unique_values\": 1,\n        \"samples\": [\n          435\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Dayhoff_for_original_aa\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 1,\n        \"samples\": [\n          \"5\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Dayhoff_for_mutated_aa\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 1,\n        \"samples\": [\n          \"6\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"SameGroup\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 1,\n        \"samples\": [\n          \"No\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}"
            }
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# @title **Select a variant from table to further analysis**\n",
        "\n",
        "UniProt_ID = 'P49768' # @param {type:\"string\"}\n",
        "# @markdown\n",
        "# @markdown ---\n",
        "# @markdown Indicate the row number (no.) of the variant to further explore:\n",
        "row_number = 1 # @param {type:\"number\"}\n",
        "\n",
        "# Function to get values from a specific row by 'no.'\n",
        "def get_row_values(row_no):\n",
        "    row = df_variants[df_variants[\"no.\"] == row_no]\n",
        "    if not row.empty:\n",
        "        original = row[\"original_aa\"].values[0]\n",
        "        mutated = row[\"mutated_aa\"].values[0]\n",
        "        residue = row[\"residue_number\"].values[0]\n",
        "        return original, mutated, residue\n",
        "    else:\n",
        "        return None, None, None\n",
        "\n",
        "Canonical_Amino_Acid, New_Amino_Acid, Variant_Position = get_row_values(row_number)\n",
        "\n",
        "# Function to format amino acid\n",
        "def format_aa(aa):\n",
        "    aa_3to1 = {\n",
        "        \"ALA\": \"A\", \"ARG\": \"R\", \"ASN\": \"N\", \"ASP\": \"D\", \"CYS\": \"C\",\n",
        "        \"GLN\": \"Q\", \"GLU\": \"E\", \"GLY\": \"G\", \"HIS\": \"H\", \"ILE\": \"I\",\n",
        "        \"LEU\": \"L\", \"LYS\": \"K\", \"MET\": \"M\", \"PHE\": \"F\", \"PRO\": \"P\",\n",
        "        \"SER\": \"S\", \"THR\": \"T\", \"TRP\": \"W\", \"TYR\": \"Y\", \"VAL\": \"V\",\n",
        "        \"A\": \"ALA (A)\", \"R\": \"ARG (R)\", \"N\": \"ASN (N)\", \"D\": \"ASP (D)\", \"C\": \"CYS (C)\",\n",
        "        \"Q\": \"GLN (Q)\", \"E\": \"GLU (E)\", \"G\": \"GLY (G)\", \"H\": \"HIS (H)\", \"I\": \"ILE (I)\",\n",
        "        \"L\": \"LEU (L)\", \"K\": \"LYS (K)\", \"M\": \"MET (M)\", \"F\": \"PHE (F)\", \"P\": \"PRO (P)\",\n",
        "        \"S\": \"SER (S)\", \"T\": \"THR (T)\", \"W\": \"TRP (W)\", \"Y\": \"TYR (Y)\", \"V\": \"VAL (V)\"\n",
        "    }\n",
        "    one_letter = aa_3to1.get(aa)\n",
        "    return f\"{aa} ({one_letter})\" if one_letter else aa\n",
        "\n",
        "# Apply formatting\n",
        "Canonical_Amino_Acid = format_aa(Canonical_Amino_Acid)\n",
        "New_Amino_Acid = format_aa(New_Amino_Acid)\n",
        "\n",
        "# Output result\n",
        "print(f\"Original: {Canonical_Amino_Acid}, Mutated: {New_Amino_Acid}, Residue Number: {Variant_Position}\")\n",
        "\n",
        "####### COLOURS\n",
        "#defining colour variables based on amino acids identity\n",
        "\n",
        "if Canonical_Amino_Acid == 'CYS (C)' or Canonical_Amino_Acid == 'PRO (P)' or Canonical_Amino_Acid == 'GLY (G)':\n",
        "    colour1 = '#cccc00'\n",
        "    colourword1 = 'yellow'\n",
        "elif Canonical_Amino_Acid == 'SER (S)' or Canonical_Amino_Acid == 'THR (T)' or Canonical_Amino_Acid == 'ASN (N)' or Canonical_Amino_Acid == 'GLN (Q)':\n",
        "    colour1 = '#cc99ff'\n",
        "    colourword1 = 'pink-purple'\n",
        "elif Canonical_Amino_Acid == 'ARG (R)' or Canonical_Amino_Acid == 'HIS (H)' or Canonical_Amino_Acid == 'LYS (K)' or Canonical_Amino_Acid == 'ASP (D)' or Canonical_Amino_Acid == 'GLU (E)':\n",
        "    colour1 = '#9999ff'\n",
        "    colourword1 = 'blue-purple'\n",
        "elif Canonical_Amino_Acid == 'ALA (A)' or Canonical_Amino_Acid == 'VAL (V)' or Canonical_Amino_Acid == 'ILE (I)' or Canonical_Amino_Acid == 'LEU (L)':\n",
        "    colour1 = '#99ff66'\n",
        "    colourword1 = 'green'\n",
        "elif Canonical_Amino_Acid == 'MET (M)' or Canonical_Amino_Acid == 'PHE (F)' or Canonical_Amino_Acid == 'TYR (Y)' or Canonical_Amino_Acid == 'TRP (W)':\n",
        "    colour1 = '#99ff66'\n",
        "    colourword1 = 'green'\n",
        "else:\n",
        "    colour1 = '#b3b3b3'\n",
        "\n",
        "if New_Amino_Acid == 'CYS (C)' or New_Amino_Acid == 'PRO (P)' or New_Amino_Acid == 'GLY (G)':\n",
        "    colour2 = '#cccc00'\n",
        "    colourword2 = 'yellow'\n",
        "elif New_Amino_Acid == 'SER (S)' or New_Amino_Acid == 'THR (T)' or New_Amino_Acid == 'ASN (N)' or New_Amino_Acid == 'GLN (Q)':\n",
        "    colour2 = '#cc99ff'\n",
        "    colourword2 = 'pink-purple'\n",
        "elif New_Amino_Acid == 'ARG (R)' or New_Amino_Acid == 'HIS (H)' or New_Amino_Acid == 'LYS (K)' or New_Amino_Acid == 'ASP (D)' or New_Amino_Acid == 'GLU (E)':\n",
        "    colour2 = '#9999ff'\n",
        "    colourword2 = 'blue-purple'\n",
        "elif New_Amino_Acid == 'ALA (A)' or New_Amino_Acid == 'VAL (V)' or New_Amino_Acid == 'ILE (I)' or New_Amino_Acid == 'LEU (L)':\n",
        "    colour2 = '#99ff66'\n",
        "    colourword2 = 'green'\n",
        "elif New_Amino_Acid == 'MET (M)' or New_Amino_Acid == 'PHE (F)' or New_Amino_Acid == 'TYR (Y)' or New_Amino_Acid == 'TRP (W)':\n",
        "    colour2 = '#99ff66'\n",
        "    colourword2 = 'green'\n",
        "else:\n",
        "    colour2 = '#b3b3b3'\n",
        "\n",
        "if Canonical_Amino_Acid == 'CYS (C)' or Canonical_Amino_Acid == 'ASP (D)' or Canonical_Amino_Acid == 'GLU (E)' or Canonical_Amino_Acid == 'THR (T)':\n",
        "    colour3 = 'red'\n",
        "    colourword3 = 'red'\n",
        "elif Canonical_Amino_Acid == 'SER (S)' or Canonical_Amino_Acid == 'TYR (Y)':\n",
        "    colour3 = '#ff6666'\n",
        "    colourword3 = 'light red'\n",
        "elif Canonical_Amino_Acid == 'ARG (R)' or Canonical_Amino_Acid == 'LYS (K)':\n",
        "    colour3 = 'blue'\n",
        "    colourword3 = 'blue'\n",
        "elif  Canonical_Amino_Acid == 'HIS (H)' or Canonical_Amino_Acid == 'TRP (W)' or Canonical_Amino_Acid == 'ASN (N)' or Canonical_Amino_Acid == 'GLN (Q)':\n",
        "    colour3 = '#6699ff'\n",
        "    colourword3 = 'light blue'\n",
        "elif Canonical_Amino_Acid == 'ALA (A)' or Canonical_Amino_Acid == 'VAL (V)' or Canonical_Amino_Acid == 'ILE (I)' or Canonical_Amino_Acid == 'LEU (L)':\n",
        "    colour3 = '#b3b3b3'\n",
        "    colourword3 = 'grey'\n",
        "elif Canonical_Amino_Acid == 'MET (M)' or Canonical_Amino_Acid == 'PHE (F)' or Canonical_Amino_Acid == 'PRO (P)':\n",
        "    colour3 = '#b3b3b3'\n",
        "    colourword3 = 'grey'\n",
        "else:\n",
        "    colour3 = '#b3b3b3'\n",
        "    colourword3 = 'grey'\n",
        "\n",
        "if New_Amino_Acid == 'CYS (C)' or New_Amino_Acid == 'ASP (D)' or New_Amino_Acid == 'GLU (E)' or New_Amino_Acid == 'THR (T)':\n",
        "    colour4 = 'red'\n",
        "    colourword4 = 'red'\n",
        "elif New_Amino_Acid == 'SER (S)' or New_Amino_Acid == 'TYR (Y)':\n",
        "    colour4 = '#ff6666'\n",
        "    colourword4 = 'light red'\n",
        "elif New_Amino_Acid == 'ARG (R)' or New_Amino_Acid == 'LYS (K)':\n",
        "    colour4 = 'blue'\n",
        "    colourword4 = 'blue'\n",
        "elif  New_Amino_Acid == 'HIS (H)' or New_Amino_Acid == 'TRP (W)' or New_Amino_Acid == 'ASN (N)' or New_Amino_Acid == 'GLN (Q)':\n",
        "    colour4 = '#6699ff'\n",
        "    colourword4 = 'light blue'\n",
        "elif New_Amino_Acid == 'ALA (A)' or New_Amino_Acid == 'VAL (V)' or New_Amino_Acid == 'ILE (I)' or New_Amino_Acid == 'LEU (L)':\n",
        "    colour4 = '#b3b3b3'\n",
        "    colourword4 = 'grey'\n",
        "elif New_Amino_Acid == 'MET (M)' or New_Amino_Acid == 'PHE (F)' or New_Amino_Acid == 'PRO (P)':\n",
        "    colour4 = '#b3b3b3'\n",
        "    colourword4 = 'grey'\n",
        "else:\n",
        "    colour4 = '#b3b3b3'\n",
        "    colourword4 = 'grey'"
      ],
      "metadata": {
        "cellView": "form",
        "id": "f96Aqi9I76G_",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "75596e96-4625-4391-eca0-7d2adb1de184"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Original: LEU (L), Mutated: PHE (F), Residue Number: 435\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Amino acid table\n",
        "\n",
        "<img src=\"https://upload.wikimedia.org/wikipedia/commons/4/4f/ProteinogenicAminoAcids.svg\" height=\"700\" align=\"center\">"
      ],
      "metadata": {
        "id": "abDv79_773fd"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# @title Hydrophobic *vs.* Hydrophilic\n",
        "# @markdown ---\n",
        "# @markdown\n",
        "# @markdown Pressing run / shift+enter will generate a visuallization.\n",
        "# @markdown <br>A circle representing the canonical and the variant amino acid will be generated.\n",
        "\n",
        "clearscreen()\n",
        "print(f\"   canonical ({colourword1})  -->   variant({colourword2})\")\n",
        "setup(500,125)\n",
        "T = Turtle()\n",
        "T.color(colour1)\n",
        "T.speed(15)\n",
        "S = T.clone()\n",
        "S.color(colour2)\n",
        "T.jumpto(-150,50)\n",
        "S.jumpto(25,50)\n",
        "T.begin_fill()\n",
        "S.begin_fill()\n",
        "T.circle(-50)\n",
        "S.circle(-50)\n",
        "T.end_fill()\n",
        "S.end_fill()\n",
        "T.jumpto(-150,25)\n",
        "S.jumpto(25,25)\n",
        "\n",
        "print(\"      (comparing amino acid hydrophobocity)\")\n"
      ],
      "metadata": {
        "cellView": "form",
        "id": "BfMKz-505urS",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 182
        },
        "outputId": "134ad467-9c49-43fa-cbc2-2ea38a68218e"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": [
              "\n",
              "      <svg width=\"500\" height=\"125\">  \n",
              "        <rect width=\"100%\" height=\"100%\" style=\"fill:white;stroke:;stroke-width:1\"/>\n",
              "        \n",
              "        \n",
              "        <path fill-rule=\"evenodd\" fill-opacity=\"1\" d=\"M 100.0 12.5  A -50 -50 0 0 1 100.0 112.5  A -50 -50 0 0 1 100.0 12.5 \" stroke-linecap=\"round\" style=\"stroke:#99ff66;stroke-width:1\" fill=\"#99ff66\" /><path fill-rule=\"evenodd\" fill-opacity=\"1\" d=\"M 275.0 12.5  A -50 -50 0 0 1 275.0 112.5  A -50 -50 0 0 1 275.0 12.5 \" stroke-linecap=\"round\" style=\"stroke:#99ff66;stroke-width:1\" fill=\"#99ff66\" />\n",
              "        \n",
              "        \n",
              "        <g id=\"classic\" visibility=\"visible\" transform=\"rotate(-90.0,100.0,37.5) translate(100.0, 37.5)\">\n",
              "<polygon points=\"-5,-4.5 0,-2.5 5,-4.5 0,4.5\" transform=\"skewX(0) scale(1,1)\" style=\"stroke:#99ff66;fill:#99ff66;stroke-width:1\" />\n",
              "</g><g id=\"classic\" visibility=\"visible\" transform=\"rotate(270.0,275.0,37.5) translate(275.0, 37.5)\">\n",
              "<polygon points=\"-5,-4.5 0,-2.5 5,-4.5 0,4.5\" transform=\"skewX(0.0) scale(1,1)\" style=\"stroke:#99ff66;fill:#99ff66;stroke-width:1\" />\n",
              "</g>\n",
              "      </svg>\n",
              "    "
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "   canonical (green)  -->   variant(green)\n",
            "      (comparing amino acid hydrophobocity)\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "The side chains on amino acids can be classed as follows:\n",
        "\n",
        "* **<font color='#9999ff'> Polar charged side chains</font>** (hydrophilic; blue-purple)\n",
        "* **<font color='#cc99ff'> Polar uncharged side chains</font>** (hydrophilic; pink-purple)\n",
        "* **<font color='#99ff66'> Hydrophobic & aliphatic side chains</font>** (hydrophobic; green)\n",
        "* **<font color='#99ff66'> Hydrophobic & aromatic side chains</font>** (hydrophobic: green)\n",
        "* **<font color='#cccc00'> Special case side chains</font>** (yellow)\n",
        "\n",
        "The above colour contrast shows if the variant changes hydrophobicity."
      ],
      "metadata": {
        "id": "HJBBsOD5BfWw"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# @title Acidic *vs.* Basic\n",
        "# @markdown ---\n",
        "# @markdown Pressing run / shift+enter will generate a visuallization.\n",
        "# @markdown <br>A circle representing the canonical and the variant amino acid will be generated.\n",
        "\n",
        "clearscreen()\n",
        "print(f\"      canonical ({colourword3})  -->  variant ({colourword4})\")\n",
        "setup(500,125)\n",
        "T = Turtle()\n",
        "T.color(colour3)\n",
        "T.speed(15)\n",
        "S = T.clone()\n",
        "S.color(colour4)\n",
        "T.jumpto(-150,50)\n",
        "S.jumpto(25,50)\n",
        "T.begin_fill()\n",
        "S.begin_fill()\n",
        "T.circle(-50)\n",
        "S.circle(-50)\n",
        "T.end_fill()\n",
        "S.end_fill()\n",
        "T.jumpto(-150,25)\n",
        "S.jumpto(25,25)\n",
        "\n",
        "print(\"      (comparing amino acid charge polarity)\")"
      ],
      "metadata": {
        "cellView": "form",
        "id": "MRK488g6OgZ6",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 182
        },
        "outputId": "c69fa46e-9400-4076-aa4d-7ea334f07a4a"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "      canonical (grey)  -->  variant (grey)\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": [
              "\n",
              "      <svg width=\"500\" height=\"125\">  \n",
              "        <rect width=\"100%\" height=\"100%\" style=\"fill:white;stroke:;stroke-width:1\"/>\n",
              "        \n",
              "        \n",
              "        <path fill-rule=\"evenodd\" fill-opacity=\"1\" d=\"M 100.0 12.5  A -50 -50 0 0 1 100.0 112.5  A -50 -50 0 0 1 100.0 12.5 \" stroke-linecap=\"round\" style=\"stroke:#b3b3b3;stroke-width:1\" fill=\"#b3b3b3\" /><path fill-rule=\"evenodd\" fill-opacity=\"1\" d=\"M 275.0 12.5  A -50 -50 0 0 1 275.0 112.5  A -50 -50 0 0 1 275.0 12.5 \" stroke-linecap=\"round\" style=\"stroke:#b3b3b3;stroke-width:1\" fill=\"#b3b3b3\" />\n",
              "        \n",
              "        \n",
              "        <g id=\"classic\" visibility=\"visible\" transform=\"rotate(-90.0,100.0,37.5) translate(100.0, 37.5)\">\n",
              "<polygon points=\"-5,-4.5 0,-2.5 5,-4.5 0,4.5\" transform=\"skewX(0) scale(1,1)\" style=\"stroke:#b3b3b3;fill:#b3b3b3;stroke-width:1\" />\n",
              "</g><g id=\"classic\" visibility=\"visible\" transform=\"rotate(270.0,275.0,37.5) translate(275.0, 37.5)\">\n",
              "<polygon points=\"-5,-4.5 0,-2.5 5,-4.5 0,4.5\" transform=\"skewX(0.0) scale(1,1)\" style=\"stroke:#b3b3b3;fill:#b3b3b3;stroke-width:1\" />\n",
              "</g>\n",
              "      </svg>\n",
              "    "
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "      (comparing amino acid charge polarity)\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "The above shows if the variant changes the size of the side chain\n",
        "<br>(with basic vs acidic colouring)."
      ],
      "metadata": {
        "id": "7VebV338YIX-"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "The above colour contrast shows if the variant changes the charge / polarity of the side chain:\n",
        "<br>\n",
        "\n",
        "<font color='blue'> **basic**</font> (positively charged) --> <font color='#6699ff'> **weakly basic**</font> (uncharged but polar) -->\n",
        "<br>**neutral** --> <font color='#ff6666'> **weakly acidic**</font> (uncharged but polar) --> <font color='red'> **acidic**</font> (negatively charged)\n",
        "<br>\n",
        "<br>\n",
        "<font color='red'> **acidic**</font> (negatively charged) --> <font color='#ff6666'> **weakly acidic**</font> (uncharged but polar) -->\n",
        "<br>**neutral** --> <font color='#6699ff'> **weakly basic**</font> (uncharged but polar) --> <font color='blue'> **basic**</font> (positively charged)"
      ],
      "metadata": {
        "id": "w9ZjipAQMkj7"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# @title Tiny *vs.* Small *vs.* Big\n",
        "# @markdown ---\n",
        "# @markdown Pressing run / shift+enter will generate a visuallization.\n",
        "# @markdown <br>A circle representing the canonical and the variant amino acid will be generated.\n",
        "\n",
        "clearscreen()\n",
        "print(f\"    canonical ({colourword3})  -->  variant ({colourword4})\")\n",
        "setup(500,250)\n",
        "T = Turtle()\n",
        "T.color(colour3)\n",
        "T.speed(15)\n",
        "S = T.clone()\n",
        "S.color(colour4)\n",
        "T.jumpto(-150,50)\n",
        "S.jumpto(25,50)\n",
        "T.begin_fill()\n",
        "S.begin_fill()\n",
        "\n",
        "if Canonical_Amino_Acid == 'ALA' or Canonical_Amino_Acid == 'GLY' or Canonical_Amino_Acid == 'SER':\n",
        "    T.circle(-20)\n",
        "elif Canonical_Amino_Acid == 'ASP' or Canonical_Amino_Acid == 'ASN' or Canonical_Amino_Acid == 'THR' or Canonical_Amino_Acid == 'VAL' or Canonical_Amino_Acid == 'PRO':\n",
        "    T.circle(-30)\n",
        "elif Canonical_Amino_Acid == 'ARG' or Canonical_Amino_Acid == 'HIS' or Canonical_Amino_Acid == 'PHE' or Canonical_Amino_Acid == 'TRP' or Canonical_Amino_Acid == 'TYR':\n",
        "    T.circle(-75)\n",
        "else:\n",
        "    T.circle(-50)\n",
        "\n",
        "if New_Amino_Acid == 'ALA' or New_Amino_Acid == 'GLY' or New_Amino_Acid == 'SER':\n",
        "    S.circle(-20)\n",
        "elif New_Amino_Acid == 'ASP' or New_Amino_Acid == 'ASN' or New_Amino_Acid == 'THR' or New_Amino_Acid == 'VAL' or New_Amino_Acid == 'PRO':\n",
        "    S.circle(-30)\n",
        "elif New_Amino_Acid == 'ARG' or New_Amino_Acid == 'HIS' or New_Amino_Acid == 'PHE' or New_Amino_Acid == 'TRP' or New_Amino_Acid == 'TYR':\n",
        "    S.circle(-75)\n",
        "else:\n",
        "    S.circle(-50)\n",
        "\n",
        "T.end_fill()\n",
        "S.end_fill()\n",
        "T.jumpto(-150,25)\n",
        "S.jumpto(25,25)\n",
        "\n",
        "print(\"      (comparing amino acid charge steric bulk)\")\n"
      ],
      "metadata": {
        "cellView": "form",
        "id": "6DO8jooRYHZv",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 307
        },
        "outputId": "395231d4-40f2-46ae-eef0-78dd7822524b"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "    canonical (grey)  -->  variant (grey)\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": [
              "\n",
              "      <svg width=\"500\" height=\"250\">  \n",
              "        <rect width=\"100%\" height=\"100%\" style=\"fill:white;stroke:;stroke-width:1\"/>\n",
              "        \n",
              "        \n",
              "        <path fill-rule=\"evenodd\" fill-opacity=\"1\" d=\"M 100.0 75.0  A -50 -50 0 0 1 100.0 175.0  A -50 -50 0 0 1 100.0 75.0 \" stroke-linecap=\"round\" style=\"stroke:#b3b3b3;stroke-width:1\" fill=\"#b3b3b3\" /><path fill-rule=\"evenodd\" fill-opacity=\"1\" d=\"M 275.0 75.0  A -50 -50 0 0 1 275.0 175.0  A -50 -50 0 0 1 275.0 75.0 \" stroke-linecap=\"round\" style=\"stroke:#b3b3b3;stroke-width:1\" fill=\"#b3b3b3\" />\n",
              "        \n",
              "        \n",
              "        <g id=\"classic\" visibility=\"visible\" transform=\"rotate(-90.0,100.0,100.0) translate(100.0, 100.0)\">\n",
              "<polygon points=\"-5,-4.5 0,-2.5 5,-4.5 0,4.5\" transform=\"skewX(0) scale(1,1)\" style=\"stroke:#b3b3b3;fill:#b3b3b3;stroke-width:1\" />\n",
              "</g><g id=\"classic\" visibility=\"visible\" transform=\"rotate(270.0,275.0,100.0) translate(275.0, 100.0)\">\n",
              "<polygon points=\"-5,-4.5 0,-2.5 5,-4.5 0,4.5\" transform=\"skewX(0.0) scale(1,1)\" style=\"stroke:#b3b3b3;fill:#b3b3b3;stroke-width:1\" />\n",
              "</g>\n",
              "      </svg>\n",
              "    "
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "      (comparing amino acid charge steric bulk)\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Local structural environment of the variant position"
      ],
      "metadata": {
        "id": "WTDelvgecjFc"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "## **Buried?**\n",
        "\n",
        "<img src=\"https://upload.wikimedia.org/wikipedia/commons/c/c5/Protein_folding_schematic.png\" height=\"500\" align=\"right\">\n",
        "\n",
        "<p>Amino acids can be positioned at the protein surface or buried inside a structural fold.</p>\n",
        "\n",
        "<br>\n",
        "<p>Often hydrophobic ('water-hating') amino acids are buried, and hydrophilic ('water-loving') amino acids are on the protein surface. The image on the right shows almost all hydrophobic residues (black) buried and all hydrophilic residues (white) on the surface. This is a simplification because in most structures a mixture of amino acid types is observed buried within the structural fold. However, the buried hydrophilic residues tend to have amino acid partners that are forming complementary  interactions (salt-bridges, hydrogen bonds, etc).\n",
        "</p>\n",
        "\n",
        "<br>\n",
        "<p>If an amino acid appears at the surface when considering a single protein chain, it may be buried by protein-protein interactions when considering a higher-level complex, that is the protein's quaternary structure. For example, an amino acid on the surface when considering a single chain, may be buried between protein chains within a protein, or by a protein partner that forms a temporary complex with another protein or protein(s) for a specific purpose.</p>\n",
        "\n",
        "<br>\n",
        "<p>We will first only be considering the amino acid at the variant position within the context of a single protein chain.\n",
        "</p>"
      ],
      "metadata": {
        "id": "MVQ4_xq20W6l"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# @title Question\n",
        "# @markdown\n",
        "# @markdown\n",
        "# @markdown ---\n",
        "# @markdown Where is the **canonical** amino acid with respect to the rest of the single protein chain?\n",
        "FOUND_AT = 'Buried_site' # @param [\"Input Answer\",\"Protein_surface\", \"Buried_site\",\"Unclear\"]\n",
        "# @markdown **[PDF Guide for using Mol* to answer this question](https://github.com/PDBeurope/pdbe-notebooks/blob/main/variants_embl-ebi_may2024/Molstar_Guide_for_Buried_or_Surface_Question.pdf)**\n",
        "# @markdown <br> HINT1: Look at sequence viewer on **PDBe-KB** Structures page. Take note of which experimentally structures contain the sequence region with the variant position.\n",
        "# @markdown <br>HINT2: Use structural superpositions on **PDBe-KB** Structures page and/or view a protein structure / structure(s) using [Mol* viewer](https://molstar.org/viewer/) to aid analysis\n",
        "# @markdown <br>HINT3: In Mol* viewer, click on triple dots next to polymer and add **'Molecular surface'** or **'Spacefill'** representation.\n",
        "# @markdown <br>HINT4: Use sequence view in Mol* and note the UniProt ID sequence numbering that appears in the bottom right corner.\n",
        "# @markdown\n",
        "# @markdown ---"
      ],
      "metadata": {
        "cellView": "form",
        "id": "0ljyJfqt0VKl"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Salt bridges & hydrogen bonds\n",
        "<img src=\"https://upload.wikimedia.org/wikipedia/commons/b/b4/Next_Revisit_Glutamic_Acid_Lysine_salt_bridge.png\" height=\"300\" align=\"right\">\n",
        "\n",
        "A key feature contributing to protein folding and protein stability are hydrogen bonds and salt bridges. A salt bridge is a combination of hydrogen bonding and ionic bonding.\n",
        "\n",
        "<br>\n",
        "\n",
        "The most common amino acid pairs involved in a salt bridge:\n",
        "\n",
        "*  ASP - LYS\n",
        "*  ASP - ARG\n",
        "*  GLU - LYS\n",
        "*  GLU - ARG\n",
        "\n",
        "<br>\n",
        "\n",
        "If an acidic or basic amino acid is buried within a structural fold, it is typically forming either a salt bridge or hydrogen bonds with another amino acid.\n"
      ],
      "metadata": {
        "id": "ZzyyJytjBTty"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# @title Question\n",
        "# @markdown\n",
        "# @markdown\n",
        "# @markdown ---\n",
        "# @markdown If the **canonical** amino acid is basic or acidic, as well as buried, <br> is there an amino acid partner that can be identified?\n",
        "ACIDIC_or_BASIC_PARTNER = 'No' # @param [\"Input Answer\",\"Yes\", \"No\",\"Unclear\",\"Canonical_aa_not_acidic_or_basic\"]\n",
        "# @markdown HINT: Use structural superpositions on **PDBe-KB** Structures page and/or view a protein structure / structure(s) using mol* to aid analysis\n",
        "# @markdown\n",
        "# @markdown ---"
      ],
      "metadata": {
        "cellView": "form",
        "id": "GfL0alHB1CBv"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Disulfide bridges & metal coordination\n",
        "<img src=\"https://upload.wikimedia.org/wikipedia/commons/8/8c/Disulfide_Bridges_(SCHEMATIC)_V.1.svg\" height=\"250\" align=\"right\">\n",
        "\n",
        "<p>Key features that contribute to protein folding and stability are disulfide bridges.  The image on the right highlights how the disulfide bridges act as covalent crosslinkers to stabilise a protein's fold. Metal coordination has also been observed as important for protein folding and stability.</p>\n",
        "\n",
        "<br>\n",
        "<p>Disulfide bonds and metal coordination may also contribute to catalysis when the protein is an enzyme. This may serve as an additional role to protein stabilisation. </p<>\n",
        "\n",
        "<br>\n",
        "<br>\n",
        "<p>Disulfide bonds only involve cysteines, and metal coordination tends to involve cysteines and histidines, but can involve other amino acids.</p>\n"
      ],
      "metadata": {
        "id": "WAuim1BRDW_z"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# @title Questions\n",
        "# @markdown\n",
        "# @markdown\n",
        "# @markdown HINT: Answering these questions can be aided by utilissing the structural superposition view and sequence viewer available from **PDBe-KB** Ligands page.\n",
        "# @markdown\n",
        "# @markdown ---\n",
        "# @markdown Is the **canonical** (reference / wild-type) amino acid a cysteine?\n",
        "Is_CYS = 'No' # @param [\"Input Answer\",\"Yes\", \"No\"]\n",
        "# @markdown ---\n",
        "# @markdown If the answer for the previous is **yes**, <br>is the cysteine involved in a disulfide bridge?\n",
        "Disulfide_bridge = 'No' # @param [\"Input Answer\",\"Yes\", \"No\"]\n",
        "# @markdown ---\n",
        "# @markdown If the answer for the previous is **no**, <br>is the cysteine involved in coordinating a metal or other ion?\n",
        "Metal_site_with_CYS = 'No' # @param [\"Input Answer\",\"Yes\", \"No\"]\n",
        "# @markdown ---\n",
        "# @markdown Is the **canonical** amino acid a histidine?\n",
        "Is_HIS = 'No' # @param [\"Input Answer\",\"Yes\", \"No\"]\n",
        "# @markdown ---\n",
        "# @markdown If the answer for the previous is **yes**, <br>is the histidine involved in coordinating a metal or other ion?\n",
        "Metal_site_with_His = 'No' # @param [\"Input Answer\",\"Yes\", \"No\"]\n",
        "# @markdown ---\n",
        "# @markdown Examine the structural superposition on the **PDBe-KB** Ligands page. <br>Are there any metals or other ions?\n",
        "Metal_or_ion_sites = 0 # @param {type:\"number\"}\n",
        "# @markdown ---\n",
        "# @markdown Are any metals involved in catalysis?\n",
        "No_of_Metals_in_catalysis = 'No' # @param [\"Input Answer\",\"Yes\", \"No\"]\n",
        "# @markdown HINT: Visit the UniProt page to see if the enzyme catalysis involves a metal.\n",
        "# @markdown\n",
        "# @markdown ---\n",
        "# @markdown If the **canonical** amino acid is involved in coordinate a metal / ion, <br>is the metal / ion site part of the active site (*i.e.* the metal / ion is involved in catalysis?)\n",
        "Metals_at_site = 'No' # @param [\"Input Answer\",\"Yes\", \"No\"]\n",
        "# @markdown ---\n",
        "\n"
      ],
      "metadata": {
        "cellView": "form",
        "id": "eV4EusrqAp10"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "##Ligand-protein interactions"
      ],
      "metadata": {
        "id": "WZ3G7tUJ0Pd5"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "<img src=\"https://github.com/PDBeurope/pdbe-notebooks/blob/main/variants_embl-ebi_may2024/Human_CFTR_ATP_Ligand.png?raw=true\" height=\"490\" align=\"right\">\n",
        "\n",
        "Small molecules (or ligands) can be found in protein structures.\n",
        "\n",
        "These can be important for protein stabilisation or function.\n",
        "\n",
        "Examples of small molecules of potential importance:\n",
        "* lipids\n",
        "* co-factors\n",
        "* carbohydrates <br>(both as substrates & as covalently attached glycosylation)\n",
        "* agonists / antagonists <br>(associated with protein receptors & channels)\n",
        "* substrates / reactants <br>(associated with protein enzymes)\n",
        "* effectors <br>(associated with protein gene regulators)\n",
        "* individual units of: <br>amino acid / DNA base / RNA base <br>\n",
        "* allosteric regulators <br>(*e.g.* amino acid acting binding to one of the enzymes involved in its synthesis)\n",
        "* inhibitors / activator\n",
        "* drugs\n",
        "\n",
        "<br>\n",
        "\n",
        "Less important small molecules*:\n",
        "* Crystallisation agents\n",
        "* Buffer components -- e.g. Tris Buffer, CC ID [TRS](https://www.ebi.ac.uk/pdbe-srv/pdbechem/chemicalCompound/show/TRS)\n",
        "* Cryoprotectants -- e.g. glycerol, CC ID [GOL](https://www.ebi.ac.uk/pdbe-srv/pdbechem/chemicalCompound/show/GOL)\n",
        "* Solvent -- e.g. DMSO, CC ID [DMS](https://www.ebi.ac.uk/pdbe-srv/pdbechem/chemicalCompound/show/DMS)\n",
        "* Detergents\n",
        "\n",
        "<p>* The second set of small molecules are those which were present for the experimental conditions utilised to determine the protein structure but are not necessarily relevant to protein function. Sometimes these agents/ additives will bind in protein pockets in ways that mimic how biologically relevant small molecules might bind, that weren't present in experimental conditions.</p>\n",
        "\n",
        "<br>\n",
        "\n",
        "---\n",
        "The **PDBe-KB** Ligands pages analyse protein structures that contain at least one protein chain with sequence mapping to a UniProt ID, but from a 'ligand' perspective.  The **PDBe-KB** Ligands page shows a list of ligands that bind the UniProt ID mapped protein chain and map the small molecule interactions onto the UniProt sequence. Additionally, there is a superposition view. Both views aim to highlight ligand binding sites and if they overlap between protein structures for the same UniProt ID. Where possible, based on chemical similarity, 'reactant-like', 'drug-like' and 'cofactor-like' chemicals have been identified.\n",
        "\n",
        "Variants that disrupt a ligand binding site that has biological relevance may impact protein function."
      ],
      "metadata": {
        "id": "5SyHXsYUBqus"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# @title Questions\n",
        "# @markdown\n",
        "# @markdown\n",
        "# @markdown HINT: Answering these questions can be aided by utilising the structural superposition view and sequence viewer available from **PDBe-KB** Ligands page.\n",
        "# @markdown\n",
        "# @markdown ---\n",
        "# @markdown Is the **canonical** amino acid near a small molecule binding site?\n",
        "Near_small_molecule = 'Yes' # @param [\"Input Answer\",\"Yes\", \"No\"]\n",
        "# @markdown ---\n",
        "# @markdown Are there any direct interactions between the **canonical** amino acid and small molecule(s)?\n",
        "Interacting_with_small_molecule = 'Yes' # @param [\"Input Answer\",\"Yes\", \"No\"]\n",
        "# @markdown ---\n",
        "# @markdown How strong is the experimental evidence supporting the small molecules binding pose?\n",
        "Map_supports_ligand_binding_pose = 'Yes' # @param [\"Input Answer\",\"Yes\", \"No\", \"Uncertain\"]\n",
        "# @markdown ---\n",
        "# @markdown Is it likely that the variant disrupts the ligand binding site?\n",
        "LIKELY_DISRUPTING_ligand_binding = 'Yes' # @param [\"Input Answer\",\"Yes\", \"No\", \"Maybe\"]\n",
        "# @markdown ---\n",
        "# @markdown ---\n",
        "# @markdown Is the small molecule at the site we are examining have a known biological function listed on the UniProt page?\n",
        "UniProt_indicates_bio_relevance = 'Yes' # @param [\"Input Answer\",\"Yes\", \"No\"]\n",
        "# @markdown ---\n",
        "# @markdown Is the small molecule at the site we are examining marked as 'co-factor-like' or 'drug factor-like' on the **PDBe-KB** Ligands page?\n",
        "PDBeKB_indicates_bio_relevance = 'No' # @param [\"Input Answer\",\"Yes\", \"No\"]\n",
        "# @markdown ---\n",
        "# @markdown Examine the structure/structures associated with the small molecule. Does the citation or protein structure 'entry title' for the protein structures indicate the small molecule has biological relevance?\n",
        "Entry_title_or_cit_indicates_bio_relevance = 'Yes' # @param [\"Input Answer\",\"Yes\", \"No\", \"Maybe\"]\n",
        "# @markdown HINT: Visit the UniProt page to see if the enzyme catalysis involves a metal.\n",
        "# @markdown\n",
        "# @markdown ---\n",
        "# @markdown Is it likely that the variant disrupting the ligand binding is disrupting protein function?\n",
        "LIKELY_DISRUPTING_function = 'Yes' # @param [\"Input Answer\",\"Yes\", \"No\", \"Maybe\"]\n",
        "\n"
      ],
      "metadata": {
        "cellView": "form",
        "id": "t1mX2JopbG3b"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Variant considered in the broader structural context"
      ],
      "metadata": {
        "id": "UGb4zrz70dja"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "## **Secondary structure**\n",
        "\n",
        "The most common secondary structure elements are $\\beta$-sheets and $\\alpha$-helices.\n",
        "\n",
        "As shown in the image below, the formation of these features involves hydrogen bonds (black dashes). The hydrogen bond network in $\\beta$-sheets or  $\\alpha$-helices involves mostly the backbone atoms (carboxyl and amino groups) of amino acids (*aka* NOT the side chain atoms that changes between amino acids and change if a genetic variant results in a new amino acid within the protein sequence). Thus, the variant is unlikely to directly impact the hydrogen bonding network, but rather the secondary structure will be indirectly impacted when the new amino acid is repositioned with respect to adjacent amino acids because it cannot be accommodated at the same position in the same way as the canonical amino acid. The exception is proline, where the backbone is fused with the side chain, so has special behaviour, if it is being replaced or being introduced by a genetic variant.\n",
        "\n",
        "<br>\n",
        "\n",
        "<img src=\"https://upload.wikimedia.org/wikipedia/commons/c/c5/Alpha_beta_structure_(full).png\" height=\"400\" align=\"center\">\n"
      ],
      "metadata": {
        "id": "YSZjjATKeMua"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# @title Questions\n",
        "# @markdown\n",
        "# @markdown\n",
        "# @markdown HINT: The following subsection can be aided by the **PDBe-KB** Structures page, especially the protein sequence view.\n",
        "# @markdown\n",
        "# @markdown ---\n",
        "# @markdown What is the secondary structure where the position of the **canonical** amino acid is?\n",
        "SECONDARY_STRUCTURE = 'Input Answer' # @param [\"Input Answer\",\"HELIX\", \"SHEET\", \"LOOP\"]\n",
        "# @markdown Selecting **LOOP** indicates NO secondary structure element is present.\n",
        "# @markdown\n",
        "# @markdown ---\n",
        "# @markdown Is there a secondary structure element in the position directly adjacent in sequence to where the position of the **canonical** amino acid?\n",
        "Adjacent_has_SS = 'Input Answer' # @param [\"Input Answer\",\"Yes\", \"No\"]\n",
        "# @markdown ---\n",
        "# @markdown If the answer to the above is **yes**, <br> what is the secondary structure element in the adjacent residue/residues to the position of the **canonical** amino acid in the protein sequence?\n",
        "ADJACENT_SECONDARY_STRUCTURE = 'Input Answer' # @param [\"Input Answer\",\"HELIX\", \"SHEET\", \"LOOP\"]\n",
        "# @markdown ---\n",
        "# @markdown Is it likely that the variant disrupts the secondary structure?\n",
        "LIKELY_DISRUPTING_SS= 'Input Answer' # @param [\"Input Answer\",\"Yes\", \"No\", \"Maybe\"]\n",
        "# @markdown <br>\n",
        "# @markdown Some examples that might disrupt secondary structure are:\n",
        "# @markdown\n",
        "# @markdown   *  any changes that involve a proline\n",
        "# @markdown   *  *potentially, if an amino acid is at a buried site,* and the canonical amino acid is small and the new amino acid is large\n",
        "# @markdown   *  *potentially, if an amino acid is acidic / basic,* and it changes charge thus no longer able to form salt-bridge or hydrogen bonds with appropriate amino acid partner(s).\n",
        "\n",
        "# @markdown ---\n",
        "# @markdown\n",
        "# @markdown It should be noted that disrupting secondary structure will NOT necessarily impact on protein function.\n"
      ],
      "metadata": {
        "cellView": "form",
        "id": "1IPCB6GddG62"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## **Protein domains**\n",
        "\n",
        "<H3>A protein domain can be thought of as a combination of secondary elements, folded in a compact manner, that is found amongst diverse proteins (<i>i.e.</i> found in proteins with different functions or from different organisms). It is typically thought to be a section of protein sequence which folds independently of other sections when the protein is being translated from RNA.</H3>\n",
        "\n",
        "<p>However, it is not necessary that a function is contained within a single domain -- <i>e.g.</i> an enzyme's active site may be at the interface between two domains. Thus a 'feature' / 'functional unit', or 'protein family' such as might be highlighted on UniProt pages, may or may not correspond to a protein domain.</p>\n",
        "\n",
        "Protein domains have generally been defined / understood in the context of multiple sequences and structural comparisons such as that available from databases like [InterPro](https://www.ebi.ac.uk/interpro/), [Pfam](http://pfam.xfam.org/), [SCOPe](https://scop.berkeley.edu/), SCOP, [CATH](https://www.cathdb.info/), etc). Protein classifications such as families / domains that are defined by these resources may or may not overlap for the same protein. For example, a Pfam annotation of a protein family, may correspond to two or more domains, as defined in SCOPe, SCOP or CATH. A quick overview of the type of classification / analysis performed by these databases is available [here](https://www.ebi.ac.uk/interpro/about/interpro/)<br>\n",
        "\n",
        "<br>\n",
        "\n",
        "---\n",
        "\n",
        "For even more details:\n",
        "[EMBL-EBI on-demand training course on protein classification](https://www.ebi.ac.uk/training/online/courses/protein-classification-intro-ebi-resources/protein-classification/)\n",
        "\n",
        "---\n",
        "\n",
        "<p>Investigating domains and finding related proteins with the same / similar domains may be an avenue of further analysis. Especially if no rationale can be developed to understand why a variant is causing disease after answering all the questions in this guide. Additionally, if this guide reveals a rationale for understanding why a variant causes disease, related proteins may provide further evidence or insight to support the answer.</p>\n",
        "\n",
        "---\n",
        "\n",
        "<br>"
      ],
      "metadata": {
        "id": "Qh16G240BUL9"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "A valuable tool for evaluating structures is the **Predicted aligned error (PAE)** plots associated with predicted protein structures, such as those available from the [AlphaFold database](https://alphafold.ebi.ac.uk/). The **PAE** plot is a relationship matrix that indicates confidence with respect to two amino acids positions relative to each other.\n",
        "\n",
        "<p>Thus, perhaps unsurprisingly, blocks of green, showing confidence in the prediction for the relative position of amino acids to each other often correlates with protein domains, but not always. That is, sometimes a block of green corresponds to two or more domains.</p>\n",
        "\n",
        "<br>\n",
        "\n",
        "---\n",
        "\n",
        "One example of this type of occurrence (two domains corresponding to one green square in a **PAE** plot) is pyruvate kinase PKM.<br>\n",
        "To explore this further:<br>\n",
        "[PDBe-KB page for pyruvate kinase PKM](https://www.ebi.ac.uk/pdbe/pdbe-kb/proteins/P14618)\n",
        "<br>\n",
        "[AlphaFold Database page for pyruvate kinase PKM](https://alphafold.ebi.ac.uk/entry/P14618)\n",
        "<br>\n",
        "[Example PDBe page for one of the pyruvate kinase PKM structures](https://www.ebi.ac.uk/pdbe/entry/pdb/1t5a)<br>\n",
        "[C-term domain on InterPro](https://www.ebi.ac.uk/interpro/entry/InterPro/IPR036918/)<br>\n",
        "[C-term domain on CATH](http://www.cathdb.info/version/latest/superfamily/3.40.1380.20/superposition)\n",
        "\n",
        "---\n",
        "\n",
        "<br>\n",
        "\n",
        "<p>In the next subsection we will examine the <strong>PAE</strong> plot from the <strong>AlphaFold Database</strong>, as well as structural superpositions from <strong>PDBe-KB</strong> and consider the experimental structures in the context of the full UniProt sequence.</p>\n",
        "\n",
        "<p>When considering the full protein sequence (<i>i.e.</i> UniProt ID sequence) it can appear there are multiple globular proteins connected by relatively unstructured regions of the protein chain. Where the variant occurs within the sequence has relevance to how it may or may not impair protein function.</p>\n",
        "\n",
        "<p>We will avoid the term domains in the context of the next section of analysis. This is because even though the green square on a <strong>PAE</strong> plot or the section of protein that behaved well enough to determine a structure experimentally may correspond to a single domain, it may also correspond to two or more domains. Thus, we will use the term 'folded unit'.</p>"
      ],
      "metadata": {
        "id": "WoUMAh-DE-Yf"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "\n",
        "\n",
        "---\n",
        "\n",
        "<br>\n",
        "\n",
        "Please also consider the protein processing that occurs after protein is translated.\n",
        "\n",
        "Signal peptides and propeptides are associated with 'molecular processing' events that can occur and are listed in the Feature viewer on the corresponding UniProt page or **PDBe-KB** Structure page. These are regions removed at various stages and thus are typically not present in the mature form of the protein.\n",
        "\n",
        "Sometimes multiple protein chains are generated from a single UniProt ID. This is due to specific additional molecular processing events that occur after the protein has been translated. When this has been annotated by UniProt, two or more chains with special subpages are shown on the **PDBe-KB** Summary page."
      ],
      "metadata": {
        "id": "dcTzXd0NcxtZ"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# @title Questions\n",
        "# @markdown\n",
        "# @markdown\n",
        "# @markdown HINT: The following subsection can be aided by utilising the structural superpositions available from the **PDBe-KB** Structures page,<br> especially the **AlphaFold Superposition** feature.\n",
        "# @markdown <br>HINT2: Visiting the **[AlphaFold Database](https://alphafold.ebi.ac.uk/)** provides an interactive **PAE** plot.\n",
        "# @markdown\n",
        "# @markdown ---\n",
        "# @markdown Does the **PAE** plot for the predicted structure from the **AlphaFold Database** for this UniProt ID indicate there is more than one 'folded unit'?\n",
        "More_than_one_folded_unit = 'Input Answer' # @param [\"Input Answer\",\"Yes\", \"No\", \"Maybe\"]\n",
        "# @markdown ---\n",
        "# @markdown Do the experimentally-determined structures for the UniProt ID cover different regions of sequence?\n",
        "BlueBlocks_Sequence = 'Input Answer' # @param [\"Input Answer\",\"Yes\", \"No\"]\n",
        "# @markdown ---\n",
        "# @markdown Examine the structural superposition. <br> Is there more than one superposition for the same UniProt ID?\n",
        "More_than_one_section_for_Superposition ='Input Answer' # @param [\"Input Answer\",\"Yes\", \"No\"]\n",
        "# @markdown HINT: Click on 'Select Segment' in **PDBe-KB** structural superposition view.\n",
        "# @markdown\n",
        "# @markdown ---\n",
        "# @markdown Does the protein adopt a single compact fold / appear to be a single approximately spherical shape (aka globular protein)?\n",
        "One_folded_unit = 'Input Answer' # @param [\"Input Answer\",\"Yes\", \"No\"]\n",
        "# @markdown ---\n",
        "# @markdown If there are molecular processing events that generate more than one chain, please indicate the number of chains:\n",
        "Molecular_Processing_Sections = 0 # @param {type:\"number\"}\n",
        "# @markdown HINT: Molecule processing is on **PDBe-KB** Overview page, as well as on the UniProt page *Feature viewer*\n",
        "# @markdown\n",
        "# @markdown ---\n",
        "# @markdown Taking into account any molecular processing events, has the analysis indicated there are more than one 'folded units'?\n",
        "Multiple_folded_units = 'Input Answer' # @param [\"Input Answer\",\"Yes\", \"No\"]\n",
        "# @markdown ---\n",
        "# @markdown If the answer for the previous is **yes**, <br> is the variant in the folded unit along the protein sequence, or is it part of folded unit 2, or another?\n",
        "Variant_in_folded_unit_no = 0 # @param {type:\"number\"}\n",
        "# @markdown ---\n",
        "# @markdown **Remember the above analysis for the next section.**\n",
        "# @markdown <br>**Focus the subsequent analysis on protein complexes on the folded unit which contains the canonical amino acid and ignore complexes for other folded units.**"
      ],
      "metadata": {
        "cellView": "form",
        "id": "N1JhKutMDWiL"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Protein complexes\n",
        "<img src=\"https://github.com/PDBeurope/pdbe-notebooks/blob/main/variants_embl-ebi_may2024/TB_DHDPS_Assemblies.png?raw=true\" height=\"690\" align=\"right\">\n",
        "\n",
        "<p>We have been primarily considering protein structures within the context of a single protein chain, however many proteins form complexes and many have a biological assembly of two (dimers), three (trimers), four (tetramers), five (pentamers), six (hexamers) etc</p>\n",
        "\n",
        "<br>\n",
        "\n",
        "Oligomers of protein can be broadly classified as:\n",
        "* homomeric - composed of multiple chains with the same amino acid sequence.\n",
        "* heteromeric - composed of multiple chains with different amino acid sequences.\n",
        "\n",
        "<br>\n",
        "\n",
        "If the variant is at the surface or on a secondary structure element that if modified could impact on the shape of the surface, the biological assembly of a protein could be impacted, and this could impact on the protein's function.\n",
        "\n",
        "<br>\n",
        "\n",
        "The example on the right is of an engineered mutation that disrupted protein assembly of a bacterial enzyme. Because the protein was homomeric, and because of the symmetry in the protein, the single mutation from ALA to ARG, resulted in two positively charged polar amino acids positioned opposite each other at a protein-protein interaction interface. Interestingly, although enzyme function was impacted, the engineered mutation did not fully inhibit protein function and thus indicated that if this was a genetic variant, even though the protein assembly had changed, this change would not be fatal for the organism.\n",
        "\n",
        "<br>\n",
        "\n",
        "Reference: <br> [A tetrameric structure is not essential for activity in dihydrodipicolinate synthase (DHDPS) from Mycobacterium tuberculosis](https://www.sciencedirect.com/science/article/abs/pii/S000398611100186X)"
      ],
      "metadata": {
        "id": "91cSnHJ7iSzD"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# @title Questions\n",
        "# @markdown\n",
        "# @markdown ---\n",
        "# @markdown Does the protein form a complex with other chains with the same amino acid sequence?\n",
        "More_than_one = 'Input Answer' # @param [\"Input Answer\",\"Yes\", \"No\"]\n",
        "# @markdown HINT: Visit the **PDBe-KB** Interaction page & use link for **PDB** entry page for individual structures\n",
        "# @markdown\n",
        "# @markdown ---\n",
        "# @markdown Does the protein mostly appear to form a complex only with other chains of the same amino acid sequence?\n",
        "Homomeric = 'Input Answer' # @param [\"Input Answer\",\"Yes\", \"No\"]\n",
        "# @markdown HINT: Visit the **PDBe-KB** Interaction page & use link for **PDB** entry page for individual structures\n",
        "# @markdown\n",
        "# @markdown ---\n",
        "# @markdown If the answer for the previous is **yes**, <br>how many chains with the same protein sequence are involved in the biological assembly?\n",
        "Homomeric_of = 0 # @param {type:\"number\"}\n",
        "# @markdown HINT: Visit the **PDBe-KB** Interaction page & use link for **PDB** entry page for individual structures\n",
        "# @markdown\n",
        "# @markdown ---\n",
        "# @markdown If the biological assembly involves protein chains with different protein sequences, what is the total number of chains in the biological assembly?\n",
        "Heteromeric_of = 0 # @param {type:\"number\"}\n",
        "# @markdown ---\n",
        "# @markdown If the **canonical** amino acid is at the surface of the single chain, <br>is it buried when the protein forms its biological assembly?\n",
        "Variant_at_biological_assembly_interface = 'Input Answer' # @param [\"Input Answer\",\"Yes\", \"No\"]\n",
        "# @markdown\n",
        "# @markdown ---\n",
        "# @markdown ---\n",
        "# @markdown Are there any experimentally-determined structures with the protein having multiple protein, DNA and/or RNA partners?\n",
        "Protein_partners = 'Input Answer' # @param [\"Input Answer\",\"Yes\", \"No\", \"Maybe\"]\n",
        "# @markdown HINT: Visit the **PDBe-KB** Interaction page & use link for **PDB** entry page for individual structures\n",
        "# @markdown\n",
        "# @markdown ---\n",
        "# @markdown If the **canonical** amino acid is at the surface of the single chain, or the biological assembly, is it buried when the protein interacts with its partner?\n",
        "Variant_at_biological_partner_interface = 'Input Answer' # @param [\"Input Answer\",\"Yes\", \"No\"]\n",
        "# @markdown\n",
        "# @markdown ---\n",
        "# @markdown ---\n",
        "# @markdown If the answer to the last question was **no** then:<br>\n",
        "# @markdown Compare the protein complexes listed on the UniProt page with the structures on the **PDBe-KB** Interaction page.\n",
        "# @markdown How many protein partners with predicted / experimental evidence for complex formation do not yet have any protein structure?\n",
        "Yet_to_have_structure_protein_complexes = 0 # @param {type:\"number\"}\n",
        "# @markdown ---\n",
        "# @markdown Is the **canonical** amino acid at the surface of the protein in all currently available experimentally-determined structures?\n",
        "Surface_aa_in_all_structure = 'Input Answer' # @param [\"Input Answer\",\"Yes\", \"No\"]\n",
        "# @markdown ---\n",
        "# @markdown If the answer to the final question is **yes**, although there is no evidence for it yet, the variant may be impacted on a protein complex where there has yet to be a structure determined.\n",
        "\n"
      ],
      "metadata": {
        "cellView": "form",
        "id": "tMQDmPE8BOmS"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# This ends the notebook form.<br>\n",
        "\n",
        "\n",
        "---\n",
        "\n",
        "<p>Please consider the evidence collected in this notebook form and decide what is the most likely rationale for why the genetic variant is disrupting function / pathogenic.</p>\n",
        "\n",
        "<p>It may not be apparent from the currently available protein structures why a genetic variant causes disease. This is also a valid conclusion.</p>\n",
        "\n",
        "<p>If there are any experimentally-determined structures containing the variant, please both examine these and look at the associated citations.\n",
        "\n",
        "<br>"
      ],
      "metadata": {
        "id": "gY1gKyyPyKf_"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "#**Summarize your hypothesis here:**\n",
        "\n",
        "The mutation is moderate in terms of changing amino acid properties (<i>e.g.</i> BLOSUM62 is 0, but the Dayhoff's property group is not the same).\n",
        "\n",
        "The amino acid at this position directly interacts with a ligand that was designed to be a transition state analog, so this variant likely directly impacts the activity of this protein / protein complex.\n",
        "\n",
        "<br>\n",
        "<br>\n",
        "\n",
        "---\n",
        "\n"
      ],
      "metadata": {
        "id": "Belgth4tFJOM"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "Copyright 2024 EMBL - European Bioinformatics Institute\n",
        "\n",
        "Licensed under the Apache License, Version 2.0 (the \"License\"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0\n",
        "\n",
        "Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an \"AS IS\" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License."
      ],
      "metadata": {
        "id": "IDW4csOzyDxq"
      }
    }
  ]
}